// @go-ui/core/dist/collection/patterns/go-gov-au-logo/go-gov-au-logo.js
import { Host, h } from "@stencil/core";
import { hasSlot, inheritAttributes } from "../../utils/helper";
/**
 * @slot crest - This should be where you reference the Australian Government crest img resource (on light background).
 * @slot crest-on-dark - This should be where you reference the Australian Government crest img resource (on dark background).
 * @slot co-brand - Add co-branding agency logo / elements here.
 */
export class GoGovAuLogo {
    constructor() {
        this.inheritedAttributes = {};
        this.height = '3rem';
        this.href = undefined;
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAttributes(this.el, ['class', 'style'], false);
    }
    render() {
        const { height, inheritedAttributes } = this;
        const Tag = this.href ? 'a' : 'div';
        return (h(Host, { key: '4ed1c716a7084ddcc2254cd88ccc9117dc34fb19' }, h(Tag, Object.assign({ key: 'c30d1901c779e570fca2cf27884515c74e33acbc' }, inheritedAttributes, { class: "logo-wrapper", style: { '--logo-height': height }, href: this.href }), h("span", { key: 'ac898d31da11022b1682755fcb5cba0c8bcfbc62', class: "main-brand light-only" }, h("slot", { key: 'fd0df3823cce5b6ea7b22d9fe37d9ea1aceb4fbe', name: "main-brand" })), h("span", { key: '9e22a32102c00ba179dff47706bc1bfc73ed46de', class: "main-brand dark-only" }, h("slot", { key: 'faea9bb3c0482a2c8ef348edccbd4eee0051144f', name: "main-brand-on-dark" })), hasSlot(this.el, 'co-brand') ? h("span", { class: "divider" }) : null, h("slot", { key: '226d5562adbdba293f0cba63c7240a008c176ab1', name: "co-brand" }))));
    }
    static get is() { return "go-gov-au-logo"; }
    static get originalStyleUrls() {
        return {
            "$": ["go-gov-au-logo.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["go-gov-au-logo.css"]
        };
    }
    static get properties() {
        return {
            "height": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Set height of the logo group. Slots will be scaled to this height.\nWidth is set to auto to avoid image distortion."
                },
                "attribute": "height",
                "reflect": false,
                "defaultValue": "'3rem'"
            },
            "href": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "If href is provided, logo will be wrapped inside an anchor (`a`) tag. By default, all attributes except `class` and `style` are passed through to the anchor tag."
                },
                "attribute": "href",
                "reflect": false
            }
        };
    }
    static get elementRef() { return "el"; }
}
