// @go-ui/core/dist/collection/components/go-blockquote/go-blockquote.js
import { h, Host } from "@stencil/core";
export class GoBlockquote {
    constructor() {
        this.citeUrl = undefined;
        this.hideQuoteMark = false;
    }
    render() {
        return (h(Host, { key: '78a03dda92ba3ff2f77b9d8c4a0f720a1ceecd96', class: { 'quote-mark': !this.hideQuoteMark } }, h("figure", { key: 'cc032a20cb02a3bf56d19f145c41745531ef886b' }, h("blockquote", { key: '24218268c913c37345cdd8f05196178b05b5919c', cite: this.citeUrl }, h("slot", { key: '4469730492c1603f68e98bd113fd9c54d63a7498' })), h("figcaption", { key: 'e26d76df13d8920f4d5e1dc2944ab12a24044786' }, h("span", { key: '042166bd73f1f1cce8fa02dc0e5c864bada57fe1' }, "\u2014"), h("slot", { key: '074bf22e966f2d17b38c5eef9b27627e19f4e25a', name: "caption" })))));
    }
    static get is() { return "go-blockquote"; }
    static get originalStyleUrls() {
        return {
            "$": ["go-blockquote.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["go-blockquote.css"]
        };
    }
    static get properties() {
        return {
            "citeUrl": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "A URL that designates a source document or message for the information quoted. This attribute is intended to point to information explaining the context or the reference for the quote.\nSee also: [MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/blockquote#attr-cite)"
                },
                "attribute": "cite-url",
                "reflect": false
            },
            "hideQuoteMark": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "hide-quote-mark",
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get elementRef() { return "el"; }
}
