// @go-ui/core/dist/collection/components/go-center/go-center.js
import { Host, h } from "@stencil/core";
export class GoCenter {
    constructor() {
        this.height = 'auto';
        this.alignText = false;
    }
    render() {
        return (h(Host, { key: 'c6eb1d2f0eb8b1ef4608ac503e7e2898ffb9f2c9', class: { 'align-text': this.alignText }, style: { height: this.height } }, h("slot", { key: '7cd556844061c677f5427d28ce855ff12b57bb78' })));
    }
    static get is() { return "go-center"; }
    static get originalStyleUrls() {
        return {
            "$": ["go-center.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["go-center.css"]
        };
    }
    static get properties() {
        return {
            "height": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Height of this container"
                },
                "attribute": "height",
                "reflect": false,
                "defaultValue": "'auto'"
            },
            "alignText": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If true, the text-align: center will also be applied to the container"
                },
                "attribute": "align-text",
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get elementRef() { return "el"; }
}
