// @go-ui/core/dist/collection/patterns/go-heading-row/go-heading-row.js
import { Host, h } from "@stencil/core";
export class GoHeadingRow {
    constructor() {
        this.heading = undefined;
        this.headingTag = 'h2';
        this.moreLinkHref = undefined;
        this.moreLinkText = undefined;
    }
    componentWillLoad() {
        this.hasHeadingSlot = !!this.el.querySelector('slot[name="heading"]');
        this.hasActionSlot = !!this.el.querySelector('slot[name="action"]');
    }
    render() {
        const { heading, moreLinkHref, moreLinkText, hasHeadingSlot, hasActionSlot, headingTag: HeadingTag } = this;
        const shouldRenderHeading = heading || hasHeadingSlot;
        const shouldRenderMoreLink = (moreLinkHref && moreLinkText) || hasActionSlot;
        return (h(Host, { key: '5e2070f28506c8f26bb35f5b7aacb7b4680dea05' }, shouldRenderHeading || shouldRenderMoreLink ? (h("div", { class: "heading-row" }, shouldRenderHeading ? (h("slot", { name: "heading" }, h(HeadingTag, null, heading))) : (h("span", null)), shouldRenderMoreLink ? (h("slot", { name: "action" }, h("go-nav-link", { item: {
                url: moreLinkHref,
                label: moreLinkText,
            }, showArrow: true }))) : null)) : null));
    }
    static get is() { return "go-heading-row"; }
    static get originalStyleUrls() {
        return {
            "$": ["go-heading-row.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["go-heading-row.css"]
        };
    }
    static get properties() {
        return {
            "heading": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Heading for this card row section"
                },
                "attribute": "heading",
                "reflect": false
            },
            "headingTag": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "HTML tag to use for heading"
                },
                "attribute": "heading-tag",
                "reflect": false,
                "defaultValue": "'h2'"
            },
            "moreLinkHref": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "View more link href"
                },
                "attribute": "more-link-href",
                "reflect": false
            },
            "moreLinkText": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "View more link text"
                },
                "attribute": "more-link-text",
                "reflect": false
            }
        };
    }
    static get elementRef() { return "el"; }
}
