// @go-ui/core/dist/components/go-badge.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';

const goBadgeCss = "go-badge{--badge-font-size:var(--go-size--1);--badge-size:2em;--badge-bg-color:var(--go-color-critical-600);--badge-fg-color:var(--go-color-lightest);--badge-dot-size:0.75rem;position:relative;display:inline-block}go-badge .badge-count{position:absolute;top:0;right:0;transform:translate(50%, -50%);font-size:var(--badge-font-size);font-variant-numeric:tabular-nums;text-align:center;min-width:var(--badge-size);min-height:var(--badge-size);padding:0 0.5rem;height:var(--badge-size);border-radius:var(--radius-round);color:var(--badge-fg-color);background-color:var(--badge-bg-color);font-weight:700;display:flex;align-items:center;justify-content:center;line-height:1}go-badge .badge-count.dot-only{min-width:var(--badge-dot-size);min-height:var(--badge-dot-size);font-size:0;padding:0}";
const GoBadgeStyle0 = goBadgeCss;

const GoBadge$1 = /*@__PURE__*/ proxyCustomElement(class GoBadge extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.count = undefined;
        this.min = 0;
        this.max = 99;
        this.label = undefined;
        this.dotOnly = false;
    }
    render() {
        const { count, min, max, label, dotOnly } = this;
        let displayCount = `${count}`;
        if (count < min) {
            return h("slot", null);
        }
        if (count > max) {
            displayCount = `${max}+`;
        }
        return (h(Host, null, h("slot", null), h("div", { class: { 'badge-count': true, 'dot-only': dotOnly } }, label ? h("span", { class: "visually-hidden" }, label) : null, count && !dotOnly ? h("span", { "aria-hidden": "true" }, displayCount) : null)));
    }
    get el() { return this; }
    static get style() { return GoBadgeStyle0; }
}, [4, "go-badge", {
        "count": [2],
        "min": [2],
        "max": [2],
        "label": [1],
        "dotOnly": [4, "dot-only"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-badge"];
    components.forEach(tagName => { switch (tagName) {
        case "go-badge":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoBadge$1);
            }
            break;
    } });
}
defineCustomElement$1();

const GoBadge = GoBadge$1;
const defineCustomElement = defineCustomElement$1;

export { GoBadge, defineCustomElement };
