// @go-ui/core/dist/components/go-button-group.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';

const goButtonGroupCss = "go-button-group{--go-button-group-gap:1rem;display:flex;gap:var(--go-button-group-gap);flex-wrap:wrap;align-items:baseline}go-button-group.connected{flex-wrap:nowrap;gap:0}go-button-group.connected go-button{-webkit-margin-end:-0.0625rem;margin-inline-end:-0.0625rem}go-button-group.connected go-button:first-of-type .inner-button{border-top-right-radius:0;border-bottom-right-radius:0}go-button-group.connected go-button:last-of-type .inner-button{border-top-left-radius:0;border-bottom-left-radius:0}go-button-group.connected go-button:not(:first-of-type,:last-of-type){--go-btn-radius:0}";
const GoButtonGroupStyle0 = goButtonGroupCss;

const GoButtonGroup$1 = /*@__PURE__*/ proxyCustomElement(class GoButtonGroup extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.block = undefined;
        this.connected = false;
    }
    async componentWillLoad() {
        // Make buttons inside take up full width on mobile.
        if (this.block) {
            this.el.querySelectorAll('go-button').forEach((button) => {
                button.setAttribute('block', this.block);
            });
        }
    }
    render() {
        const { connected } = this;
        return (h(Host, { key: '7c61599e31e82a4168e2caf394c5f0b54da3c8e6', role: "group", class: { connected } }, h("slot", { key: 'cbec52c7432c2e52a563d51e122e6f21310471f9' })));
    }
    get el() { return this; }
    static get style() { return GoButtonGroupStyle0; }
}, [4, "go-button-group", {
        "block": [513],
        "connected": [4]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-button-group"];
    components.forEach(tagName => { switch (tagName) {
        case "go-button-group":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoButtonGroup$1);
            }
            break;
    } });
}
defineCustomElement$1();

const GoButtonGroup = GoButtonGroup$1;
const defineCustomElement = defineCustomElement$1;

export { GoButtonGroup, defineCustomElement };
