// @go-ui/core/dist/components/go-breadcrumbs2.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { p as parseJsonProp } from './helper.js';
import { d as defineCustomElement$2 } from './go-icon2.js';
import { d as defineCustomElement$1 } from './go-nav-link2.js';

const goBreadcrumbsCss = "go-breadcrumbs{--breadcrumb-item-gap:0.5rem;display:block}go-breadcrumbs go-link{--nav-link-current-bar-width:0}go-breadcrumbs ol{padding:0;margin:0;list-style:none}go-breadcrumbs ol li{align-items:center;vertical-align:middle;display:none}@media (min-width: 600px){go-breadcrumbs ol li{display:inline-flex}}go-breadcrumbs ol li:last-child{display:inline-flex;margin-right:0}go-breadcrumbs ol li:nth-last-child(2){display:inline-flex}go-breadcrumbs ol li .breadcrumb-separator{margin-left:var(--breadcrumb-item-gap);margin-right:var(--breadcrumb-item-gap);width:1rem;height:1rem;vertical-align:middle}";
const GoBreadcrumbsStyle0 = goBreadcrumbsCss;

const GoBreadcrumb = /*@__PURE__*/ proxyCustomElement(class GoBreadcrumb extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.label = 'Breadcrumbs';
        this.items = undefined;
        this.hideCurrent = false;
        this.navItems = [];
    }
    async watchItems(newItems) {
        this.navItems = this.getItems(newItems);
    }
    componentWillLoad() {
        this.navItems = this.getItems(this.items);
    }
    getItems(items) {
        const navItems = parseJsonProp(items);
        if (!this.hideCurrent) {
            return navItems;
        }
        return navItems.filter((item, i) => item.url || i !== navItems.length - 1);
    }
    render() {
        const { label, navItems, hideCurrent } = this;
        if (!(navItems === null || navItems === void 0 ? void 0 : navItems.length)) {
            return null;
        }
        return (h(Host, null, h("nav", { "aria-label": label }, h("ol", null, navItems.map((item, i) => {
            const isLast = i === navItems.length - 1;
            const isCurrentPageALink = isLast && item.url && !hideCurrent;
            return (h("li", null, h("go-nav-link", { item: item, "aria-current": isCurrentPageALink ? 'page' : null }), i < navItems.length - 1 ? (h("svg", { class: "breadcrumb-separator", xmlns: "http://www.w3.org/2000/svg", fill: "none", stroke: "currentColor", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round", viewBox: "0 0 24 24" }, h("path", { d: "m9 18 6-6-6-6" }))) : null));
        })))));
    }
    get el() { return this; }
    static get watchers() { return {
        "items": ["watchItems"]
    }; }
    static get style() { return GoBreadcrumbsStyle0; }
}, [0, "go-breadcrumbs", {
        "label": [1],
        "items": [1],
        "hideCurrent": [4, "hide-current"],
        "navItems": [32]
    }, undefined, {
        "items": ["watchItems"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-breadcrumbs", "go-icon", "go-nav-link"];
    components.forEach(tagName => { switch (tagName) {
        case "go-breadcrumbs":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoBreadcrumb);
            }
            break;
        case "go-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "go-nav-link":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}
defineCustomElement();

export { GoBreadcrumb as G, defineCustomElement as d };
