// @go-ui/core/dist/components/go-card-row.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';
import { m as moveEl } from './dom.js';
import { b as hasSlot } from './helper.js';
import { d as defineCustomElement$4 } from './go-heading-row2.js';
import { d as defineCustomElement$3 } from './go-icon2.js';
import { d as defineCustomElement$2 } from './go-nav-link2.js';

const goCardRowCss = "go-card-row{--card-row-gap:1rem;align-items:stretch;--go-col-gap:var(--card-row-gap)}go-card-row.no-stretch{align-items:flex-start}go-card-row go-card{opacity:0;max-height:12.5rem}go-card-row go-card.loaded{opacity:1;max-height:inherit}go-card-row .stagger go-card.loaded{opacity:0}go-card-row .card-col{margin-bottom:var(--card-row-gap)}";
const GoCardRowStyle0 = goCardRowCss;

const GoCardRow$1 = /*@__PURE__*/ proxyCustomElement(class GoCardRow extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        /**
         * class list for each card
         */
        this.cardClasses = ['card-col'];
        this.noStretch = false;
        this.cols = 1;
        this.colsTablet = 2;
        this.colsDesktop = 3;
        this.colsLarge = 4;
        this.stagger = undefined;
        this.heading = undefined;
        this.headingTag = 'h2';
        this.moreLinkHref = undefined;
        this.moreLinkText = undefined;
    }
    componentWillLoad() {
        // check if heading slot is used
        this.hasHeadingSlot = hasSlot(this.el, 'heading');
        // initialise intersection observer
        if (this.stagger) {
            this.inViewObserver = new IntersectionObserver((entries) => {
                entries.forEach((entry, i) => {
                    if (!entry.isIntersecting) {
                        return;
                    }
                    const card = entry.target;
                    // add stagger fade in effect
                    card.classList.add('stagger-fade-in');
                    card.style.cssText = `--stagger-delay: ${i * this.stagger}ms`;
                    // stop observing
                    this.inViewObserver.unobserve(card);
                });
            }, {
                threshold: 0.2, // >= 1/5th of card is in view
            });
        }
    }
    async componentDidLoad() {
        const { cols, colsTablet, colsDesktop, colsLarge } = this;
        // mobile cols
        this.cardClasses.push(`col-${Math.ceil(12 / cols)}`);
        this.cardClasses.push(`col-tablet-${Math.ceil(12 / colsTablet)}`);
        this.cardClasses.push(`col-desktop-${Math.ceil(12 / colsDesktop)}`);
        this.cardClasses.push(`col-large-${Math.ceil(12 / colsLarge)}`);
        this.loadCards();
    }
    loadCards() {
        const newCards = this.el.querySelectorAll('go-card:not(.loaded)');
        newCards.forEach((card) => {
            this.prepareCard(card, this.cardClasses);
        });
    }
    prepareCard(card, colClasses) {
        // create wrapper container
        const wrapper = document.createElement('div');
        wrapper.classList.add(...colClasses);
        // insert wrapper before el in the DOM tree
        // card.parentNode.insertBefore(wrapper, card);
        // move el into wrapper
        moveEl(card, wrapper);
        card.classList.add('loaded');
        // move card into correct element
        moveEl(wrapper, this.rowEl);
        if (this.stagger) {
            this.inViewObserver.observe(card);
        }
    }
    render() {
        const { noStretch, stagger, heading, moreLinkHref, moreLinkText, hasHeadingSlot, headingTag } = this;
        const headingRowProps = {
            heading,
            moreLinkHref,
            moreLinkText,
            headingTag,
        };
        return (h("section", { key: 'c461ee7c9d696beaf18b5d3dfd35933dd1ea3fa4' }, h("go-heading-row", Object.assign({ key: '753d97a14498115f302f49ebc3f1bef675144bf4' }, headingRowProps), hasHeadingSlot ? h("slot", { name: "heading" }) : null), h("div", { key: '8832c2812498f1ee4117a1ab103a64d3d7e20e80', class: {
                'card-row row': true,
                'no-stretch': noStretch,
                'stagger': !!stagger,
            }, ref: (el) => (this.rowEl = el) }, h("slot", { key: '825773ad5c0a4a0e79ea05358988041956f65c0e' }))));
    }
    get el() { return this; }
    static get style() { return GoCardRowStyle0; }
}, [4, "go-card-row", {
        "noStretch": [4, "no-stretch"],
        "cols": [2],
        "colsTablet": [2, "cols-tablet"],
        "colsDesktop": [2, "cols-desktop"],
        "colsLarge": [2, "cols-large"],
        "stagger": [2],
        "heading": [1],
        "headingTag": [1, "heading-tag"],
        "moreLinkHref": [1, "more-link-href"],
        "moreLinkText": [1, "more-link-text"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-card-row", "go-heading-row", "go-icon", "go-nav-link"];
    components.forEach(tagName => { switch (tagName) {
        case "go-card-row":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoCardRow$1);
            }
            break;
        case "go-heading-row":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "go-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "go-nav-link":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}
defineCustomElement$1();

const GoCardRow = GoCardRow$1;
const defineCustomElement = defineCustomElement$1;

export { GoCardRow, defineCustomElement };
