// @go-ui/core/dist/components/go-center.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';

const goCenterCss = "go-center{display:grid;width:100%;height:100%;place-items:center}go-center.align-text{text-align:center}";
const GoCenterStyle0 = goCenterCss;

const GoCenter$1 = /*@__PURE__*/ proxyCustomElement(class GoCenter extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.height = 'auto';
        this.alignText = false;
    }
    render() {
        return (h(Host, { key: 'c6eb1d2f0eb8b1ef4608ac503e7e2898ffb9f2c9', class: { 'align-text': this.alignText }, style: { height: this.height } }, h("slot", { key: '7cd556844061c677f5427d28ce855ff12b57bb78' })));
    }
    get el() { return this; }
    static get style() { return GoCenterStyle0; }
}, [4, "go-center", {
        "height": [1],
        "alignText": [4, "align-text"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-center"];
    components.forEach(tagName => { switch (tagName) {
        case "go-center":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoCenter$1);
            }
            break;
    } });
}
defineCustomElement$1();

const GoCenter = GoCenter$1;
const defineCustomElement = defineCustomElement$1;

export { GoCenter, defineCustomElement };
