// @go-ui/core/dist/components/go-dialog.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';
import { d as defineCustomElement$4 } from './go-button2.js';
import { d as defineCustomElement$3 } from './go-overlay2.js';
import { d as defineCustomElement$2 } from './go-spinner2.js';
import { u as uniqueId } from './uniqueId.js';

const goDialogCss = "go-overlay.go-dialog{--go-overlay-content-padding-y:1.5rem;--go-overlay-content-padding-x:1.5rem;--go-overlay-content-width:80%;--go-overlay-content-bg-color:var(--go-color-bg);--go-overlay-content-max-width:71.25rem;--go-overlay-content-gap:10vh;--go-overlay-close-btn-text-color:currentColor;display:grid;place-items:center}go-overlay.go-dialog .overlay-content{position:relative;height:100%;width:100%;padding:var(--go-overlay-content-padding-y) var(--go-overlay-content-padding-x);max-width:var(--go-overlay-content-max-width);background:var(--go-overlay-content-bg-color)}@media (min-width: 600px){go-overlay.go-dialog .overlay-content{width:var(--go-overlay-content-width);height:auto;margin:var(--go-overlay-content-gap) auto;border-radius:var(--radius-3)}}go-overlay.go-dialog .overlay-heading{display:flex;align-items:flex-start;justify-content:space-between}go-overlay.go-dialog .overlay-heading .heading{font-size:var(--go-size-3);font-weight:500;line-height:1.5}go-overlay.go-dialog .overlay-heading .close-btn{flex:0}";
const GoDialogStyle0 = goDialogCss;

const GoDialog$1 = /*@__PURE__*/ proxyCustomElement(class GoDialog extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        // aria-labelledby https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html
        this.headingId = uniqueId('overlay-label-');
        this.active = false;
        this.persistent = false;
        this.heading = undefined;
        this.headingTag = 'h2';
    }
    async close() {
        this.active = false;
    }
    async open() {
        this.active = true;
    }
    watchActiveHandler(isActive) {
        if (isActive) {
            this.open();
            return;
        }
        this.close();
    }
    // keep active state in sync with overlay
    handleOverlayClose() {
        this.close();
    }
    render() {
        const { active, heading, persistent, headingId, headingTag: H } = this;
        return (h("go-overlay", { key: '008b1288f599b0614f7ffa88a4a8d30116025203', class: "go-dialog", persistent: persistent, role: "dialog", "aria-modal": "true", "aria-labelledby": headingId, active: active, "aria-hidden": active ? 'false' : 'true', ref: (el) => (this.overlayEl = el), onOverlayClose: () => this.handleOverlayClose() }, h("div", { key: 'a4119d4aee0daafc73dc85df57036b9b1bb7a49c', class: "overlay-heading", id: headingId }, h("slot", { key: 'cccc2708c0c33d3ff5b37e0048d947b79ef0364e', name: "heading" }, h(H, { key: '6c202c8fb7fa309ebacb029917eab2246d780844', class: "heading" }, heading)), !persistent ? (h("go-button", { class: "close-btn", flat: true, icon: true, stack: true, variant: "text", onClick: () => this.close(), "aria-label": "Close dialog" }, h("svg", { "aria-hidden": "true", xmlns: "http://www.w3.org/2000/svg", fill: "none", stroke: "currentColor", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round", viewBox: "0 0 24 24" }, h("path", { d: "M18 6 6 18M6 6l12 12" })))) : null), h("slot", { key: 'ce5c2be7d4563350a16598925afbd784733178ee' })));
    }
    get el() { return this; }
    static get watchers() { return {
        "active": ["watchActiveHandler"]
    }; }
    static get style() { return GoDialogStyle0; }
}, [4, "go-dialog", {
        "active": [1540],
        "persistent": [4],
        "heading": [1],
        "headingTag": [1, "heading-tag"],
        "close": [64],
        "open": [64]
    }, undefined, {
        "active": ["watchActiveHandler"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-dialog", "go-button", "go-overlay", "go-spinner"];
    components.forEach(tagName => { switch (tagName) {
        case "go-dialog":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoDialog$1);
            }
            break;
        case "go-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "go-overlay":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "go-spinner":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}
defineCustomElement$1();

const GoDialog = GoDialog$1;
const defineCustomElement = defineCustomElement$1;

export { GoDialog, defineCustomElement };
