// @go-ui/core/dist/components/go-footer.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { p as parseJsonProp, w as warning, b as hasSlot } from './helper.js';
import { d as defineCustomElement$4 } from './go-icon2.js';
import { d as defineCustomElement$3 } from './go-nav-link2.js';
import { d as defineCustomElement$2 } from './go-nav-list2.js';

const goFooterCss = "go-footer{--footer-nav-section-padding:3rem 0;--footer-bottom-section-padding:2rem 0;--footer-copyright-section-padding:2rem 0;--footer-bg-color:var(--go-color-lightest);--footer-text-color:var(--go-color-text);--footer-separator-color:var(--go-color-neutral-300);display:block;background:var(--footer-bg-color);color:var(--footer-text-color)}go-footer go-nav-link{--nav-link-padding:0.5em}go-footer a,go-footer go-nav-list{--go-color-link:var(--footer-text-color);--nav-list-header-color:var(--footer-text-color)}go-footer a .nav-link-text-description,go-footer go-nav-list .nav-link-text-description{--nav-link-description-color:var(--footer-text-color)}go-footer .nav-container{padding:var(--footer-nav-section-padding)}go-footer .footer-bottom{border-top:0.0625rem solid var(--footer-separator-color);padding:var(--footer-bottom-section-padding)}go-footer .copyright{border-top:0.0625rem solid var(--footer-separator-color);padding:var(--footer-copyright-section-padding)}";
const GoFooterStyle0 = goFooterCss;

const GoFooter$1 = /*@__PURE__*/ proxyCustomElement(class GoFooter extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.hasCopyRightSlot = false;
        this.hasFooterBottomSlot = false;
        this.links = undefined;
        this.navLabel = 'Footer';
        this.navCols = 1;
        this.navColsTablet = 3;
        this.navColsDesktop = 4;
        this.listHeadingTag = 'h4';
        this.navItems = undefined;
    }
    async watchItems(newItems) {
        this.navItems = parseJsonProp(newItems);
    }
    componentWillLoad() {
        var _a;
        this.navItems = parseJsonProp(this.links);
        if (((_a = this.navItems) === null || _a === void 0 ? void 0 : _a.length) > 0 && !this.navLabel) {
            warning('Please add a unique "nav-label" in order to put navigation items into the nav landmark for better accessibility.');
        }
        // check if copyright slot is empty
        this.hasCopyRightSlot = hasSlot(this.el, 'copyright');
        this.hasFooterBottomSlot = hasSlot(this.el, 'footer-bottom');
    }
    render() {
        const { navItems, navLabel, hasCopyRightSlot, hasFooterBottomSlot, navCols, navColsTablet, navColsDesktop } = this;
        return (h(Host, { key: '80b249ff0ad0f4968770ef8d5831ddb689a03e0c' }, h("footer", { key: 'bc860de5eec9caf5554a7a0994b7d6818c8fff7b' }, navItems ? (h("div", { class: "container nav-container" }, h("nav", { "aria-label": navLabel }, h("div", { class: "row" }, navItems === null || navItems === void 0 ? void 0 : navItems.map((item) => (h("go-nav-list", { block: true, class: `col-${12 / navCols} col-tablet-${12 / navColsTablet} col-desktop-${12 / navColsDesktop}`, items: item === null || item === void 0 ? void 0 : item.children }, h("go-nav-link", { block: true, slot: "header", item: item, showArrow: true })))))))) : (h("slot", null)), hasFooterBottomSlot ? (h("div", { class: "container" }, h("div", { class: "footer-bottom" }, h("slot", { name: "footer-bottom" })))) : null, hasCopyRightSlot ? (h("div", { class: "container" }, h("div", { class: "copyright" }, h("slot", { name: "copyright" })))) : null)));
    }
    get el() { return this; }
    static get watchers() { return {
        "links": ["watchItems"]
    }; }
    static get style() { return GoFooterStyle0; }
}, [4, "go-footer", {
        "links": [1],
        "navLabel": [1, "nav-label"],
        "navCols": [2, "nav-cols"],
        "navColsTablet": [2, "nav-cols-tablet"],
        "navColsDesktop": [2, "nav-cols-desktop"],
        "listHeadingTag": [1, "list-heading-tag"],
        "navItems": [32]
    }, undefined, {
        "links": ["watchItems"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-footer", "go-icon", "go-nav-link", "go-nav-list"];
    components.forEach(tagName => { switch (tagName) {
        case "go-footer":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoFooter$1);
            }
            break;
        case "go-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "go-nav-link":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "go-nav-list":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}
defineCustomElement$1();

const GoFooter = GoFooter$1;
const defineCustomElement = defineCustomElement$1;

export { GoFooter, defineCustomElement };
