// @go-ui/core/dist/components/go-field2.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { b as hasSlot, d as initIdProps, w as warning } from './helper.js';

const goFieldCss = "go-field{display:block}";
const GoFieldStyle0 = goFieldCss;

const GoField = /*@__PURE__*/ proxyCustomElement(class GoField extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.idPrefix = undefined;
        this.controlId = undefined;
        this.labelId = undefined;
        this.prefixId = undefined;
        this.suffixId = undefined;
        this.hintId = undefined;
        this.errorId = undefined;
        this.name = undefined;
        this.label = undefined;
        this.disabled = undefined;
        this.hint = undefined;
        this.error = undefined;
        this.readonly = undefined;
        this.hasError = false;
        this.hiddenInputName = undefined;
        this.hiddenInputValue = undefined;
        this.controlElSelector = '.control';
    }
    updateErrorState() {
        this.hasError = !!this.error;
    }
    componentWillLoad() {
        this.hasIconBefore = hasSlot(this.el, 'icon-before');
        this.hasIconAfter = hasSlot(this.el, 'icon-after');
        this.hasPrefix = hasSlot(this.el, 'prefix');
        this.hasSuffix = hasSlot(this.el, 'suffix');
        this.hasHintSlot = hasSlot(this.el, 'hint');
        this.hasLabelSlot = hasSlot(this.el, 'label');
        initIdProps(this, this.el, ['label', 'prefix', 'suffix', 'hint', 'error'], this.idPrefix);
        this.updateErrorState();
    }
    componentDidLoad() {
        this.updateAttributes();
    }
    componentDidUpdate() {
        this.updateAttributes();
    }
    updateAttributes() {
        if (!this.controlEl) {
            this.controlEl = this.el.querySelector(this.controlElSelector);
            if (!this.controlEl) {
                warning(`Cannot find field control based on selector ${this.controlElSelector}. Make sure the element exists in the DOM`);
                return;
            }
        }
        const { hasPrefix, hasSuffix, prefixId, suffixId, labelId, hasHintSlot, hint, hintId, hasError, errorId } = this;
        const labelledByIds = [];
        if (hasPrefix) {
            labelledByIds.push(prefixId);
        }
        labelledByIds.push(labelId);
        if (hasSuffix) {
            labelledByIds.push(suffixId);
        }
        const describedByIds = [];
        if (hasHintSlot || hint) {
            describedByIds.push(hintId);
        }
        if (hasError) {
            describedByIds.push(errorId);
        }
        if (this.disabled) {
            this.controlEl.setAttribute('disabled', 'true');
            this.controlEl.setAttribute('aria-disabled', 'true');
        }
        else {
            this.controlEl.removeAttribute('disabled');
            this.controlEl.removeAttribute('aria-disabled');
        }
        this.controlEl.setAttribute('aria-labelledby', labelledByIds.join(' '));
        this.controlEl.setAttribute('aria-describedby', describedByIds.join(' '));
    }
    render() {
        const { controlId, label, hasLabelSlot, hint, disabled, hasError, error, readonly, hasIconAfter, hasIconBefore, hasPrefix, hasSuffix, hasHintSlot, labelId, hintId, prefixId, suffixId, errorId, hiddenInputName, hiddenInputValue, } = this;
        const showLabel = hasLabelSlot || label;
        const showHint = hasHintSlot || hint;
        const showLabelWrapper = showLabel || showHint;
        return (h(Host, { key: '829a119b4fdb8ffbd019bd7d2f552abffda51511', class: {
                'go-field': true,
                'error': hasError,
                'readonly': !!readonly,
                'disabled': !!disabled,
                'has-prefix': hasPrefix,
                'has-suffix': hasSuffix,
                'has-icon-before': hasIconBefore,
                'has-icon-after': hasIconAfter,
            } }, showLabelWrapper ? (h("div", { class: "label-wrapper" }, showLabel ? (h("label", { class: "label", htmlFor: controlId, id: labelId }, label)) : null, showHint ? (h("div", { class: "hint", id: hintId }, h("slot", { name: "hint" }, hint))) : null)) : null, h("div", { key: '9317a3a4286142d0bcb0fc2101ecbef271aeed88', class: "control-wrapper" }, hasPrefix ? (h("span", { class: "prefix presuf", "aria-hidden": "true", id: prefixId }, h("slot", { name: "prefix" }))) : null, hasIconBefore ? (h("span", { class: "control-icon icon-before" }, h("slot", { name: "icon-before" }))) : null, h("slot", { key: 'eb0c667cae9fed3096a0fcb3df858de274b7081a' }), hasSuffix ? (h("span", { class: "suffix presuf", "aria-hidden": "true", id: suffixId }, h("slot", { name: "suffix" }))) : null, readonly || hasIconAfter ? (h("span", { class: "control-icon icon-after" }, readonly ? (
        // prettier-ignore
        h("svg", { class: "lock-icon", "aria-hidden": "true", xmlns: "http://www.w3.org/2000/svg", viewBox: "0 0 24 24" }, h("g", { fill: "none" }, h("path", { d: "M0 0h24v24H0V0z" }), h("path", { d: "M0 0h24v24H0V0z", opacity: ".87" })), h("path", { d: "M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z" }))) : (h("slot", { name: "icon-after" })))) : null), hasError ? (h("div", { id: errorId, class: "error-msg" }, error)) : null, hiddenInputName ? h("input", { type: "hidden", name: hiddenInputName, value: hiddenInputValue }) : null));
    }
    get el() { return this; }
    static get watchers() { return {
        "error": ["updateErrorState"]
    }; }
    static get style() { return GoFieldStyle0; }
}, [4, "go-field", {
        "idPrefix": [1, "id-prefix"],
        "controlId": [1, "control-id"],
        "labelId": [1025, "label-id"],
        "prefixId": [1025, "prefix-id"],
        "suffixId": [1025, "suffix-id"],
        "hintId": [1025, "hint-id"],
        "errorId": [1025, "error-id"],
        "name": [1],
        "label": [1],
        "disabled": [4],
        "hint": [1],
        "error": [8],
        "readonly": [4],
        "hiddenInputName": [1, "hidden-input-name"],
        "hiddenInputValue": [1, "hidden-input-value"],
        "controlElSelector": [1, "control-el-selector"],
        "hasError": [32]
    }, undefined, {
        "error": ["updateErrorState"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-field"];
    components.forEach(tagName => { switch (tagName) {
        case "go-field":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoField);
            }
            break;
    } });
}
defineCustomElement();

export { GoField as G, defineCustomElement as d };
