// @go-ui/core/dist/components/go-gov-au-logo.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { i as inheritAttributes, b as hasSlot } from './helper.js';

const goGovAuLogoCss = "go-gov-au-logo{--divider-color:var(--go-color-darkest);--divider-width:0.0625rem;--divider-gap:1rem;--link-color:currentColor}go-gov-au-logo .logo-wrapper{text-decoration:none;display:inline-flex;flex-direction:row;align-items:center;justify-content:center;color:var(--link-color)}go-gov-au-logo .logo-wrapper .main-brand img{height:var(--logo-height, 3rem)}go-gov-au-logo .logo-wrapper .divider{width:var(--divider-width);height:auto;align-self:stretch;background-color:var(--divider-color);margin:0 var(--divider-gap)}";
const GoGovAuLogoStyle0 = goGovAuLogoCss;

const GoGovAuLogo$1 = /*@__PURE__*/ proxyCustomElement(class GoGovAuLogo extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.inheritedAttributes = {};
        this.height = '3rem';
        this.href = undefined;
    }
    componentWillLoad() {
        this.inheritedAttributes = inheritAttributes(this.el, ['class', 'style'], false);
    }
    render() {
        const { height, inheritedAttributes } = this;
        const Tag = this.href ? 'a' : 'div';
        return (h(Host, { key: '4ed1c716a7084ddcc2254cd88ccc9117dc34fb19' }, h(Tag, Object.assign({ key: 'c30d1901c779e570fca2cf27884515c74e33acbc' }, inheritedAttributes, { class: "logo-wrapper", style: { '--logo-height': height }, href: this.href }), h("span", { key: 'ac898d31da11022b1682755fcb5cba0c8bcfbc62', class: "main-brand light-only" }, h("slot", { key: 'fd0df3823cce5b6ea7b22d9fe37d9ea1aceb4fbe', name: "main-brand" })), h("span", { key: '9e22a32102c00ba179dff47706bc1bfc73ed46de', class: "main-brand dark-only" }, h("slot", { key: 'faea9bb3c0482a2c8ef348edccbd4eee0051144f', name: "main-brand-on-dark" })), hasSlot(this.el, 'co-brand') ? h("span", { class: "divider" }) : null, h("slot", { key: '226d5562adbdba293f0cba63c7240a008c176ab1', name: "co-brand" }))));
    }
    get el() { return this; }
    static get style() { return GoGovAuLogoStyle0; }
}, [4, "go-gov-au-logo", {
        "height": [1],
        "href": [1]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-gov-au-logo"];
    components.forEach(tagName => { switch (tagName) {
        case "go-gov-au-logo":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoGovAuLogo$1);
            }
            break;
    } });
}
defineCustomElement$1();

const GoGovAuLogo = GoGovAuLogo$1;
const defineCustomElement = defineCustomElement$1;

export { GoGovAuLogo, defineCustomElement };
