// @go-ui/core/dist/components/go-hero2.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { w as warning, b as hasSlot } from './helper.js';
import { d as defineCustomElement$3 } from './go-breadcrumbs2.js';
import { d as defineCustomElement$2 } from './go-icon2.js';
import { d as defineCustomElement$1 } from './go-nav-link2.js';

const goHeroCss = "go-hero{--hero-padding-top:0;--hero-padding-bottom:0;--hero-breadcrumb-gap:2rem;--hero-pre-heading-gap:0;--hero-bg-color:var(--go-color-primary-100);--hero-fg-color:var(--go-color-darkest);--hero-text-flex-basis:60%;--hero-image-flex-basis:40%;--hero-container-gap:2rem;--hero-text-bg-blur-on-img:0.625rem;--hero-text-box-padding:2rem;--hero-img-aspect-ratio:4/3;--hero-img-fit:cover;--hero-img-position:center;--hero-full-width-bg-padding:4rem 0;--hero-bg-opacity-on-img:0.5;display:block;background-color:var(--hero-bg-color);color:var(--hero-fg-color);position:relative;overflow:hidden}go-hero .full-width-bg{position:absolute;z-index:0;width:100%;top:50%;transform:translateY(-50%)}go-hero .full-width-bg>img{object-fit:cover;width:100%;height:100%;object-fit:cover}go-hero.has-full-width-bg .hero-container{padding:var(--hero-full-width-bg-padding)}go-hero.has-full-width-bg .hero-text{position:relative;-webkit-backdrop-filter:blur(var(--hero-text-bg-blur-on-img));backdrop-filter:blur(var(--hero-text-bg-blur-on-img))}go-hero.has-full-width-bg .hero-text-bg{position:absolute;top:0;left:0;width:100%;height:100%;z-index:-1;background-color:var(--hero-bg-color);opacity:var(--hero-bg-opacity-on-img)}go-hero.hide-img-on-mobile .hero-image{display:none}@media (min-width: 1024px){go-hero.hide-img-on-mobile .hero-image{display:block}}go-hero .hero-container{position:relative;z-index:1;display:flex;flex-direction:column;padding-top:var(--hero-padding-top);padding-bottom:var(--hero-padding-bottom);gap:var(--hero-container-gap)}@media (min-width: 1024px){go-hero .hero-container{flex-direction:row}}go-hero .hero-text{padding:var(--hero-text-box-padding)}go-hero .hero-text go-breadcrumbs{margin-bottom:var(--hero-breadcrumb-gap)}go-hero .hero-text .pre-heading{margin-bottom:var(--hero-pre-heading-gap)}@media (min-width: 1024px){go-hero .hero-text{flex:0 1 var(--hero-text-flex-basis)}}@media (min-width: 1024px){go-hero .hero-image{flex:0 1 var(--hero-image-flex-basis)}}go-hero .hero-image .featured-img{width:100%;height:100%;object-fit:var(--hero-img-fit);object-position:var(--hero-img-position);aspect-ratio:var(--hero-img-aspect-ratio)}";
const GoHeroStyle0 = goHeroCss;

const GoHero = /*@__PURE__*/ proxyCustomElement(class GoHero extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.hasFullWidthBg = false;
        this.heading = undefined;
        this.preHeading = undefined;
        this.breadcrumbs = undefined;
        this.imgSrc = undefined;
        this.imgAlt = undefined;
        this.hideImgOnMobile = false;
    }
    watchImgSrc(value) {
        if (value && !this.imgAlt) {
            warning('img-alt attribute is required when img-src is present');
        }
    }
    watchImgAlt(value) {
        if (value && !this.imgSrc) {
            warning('img-src attribute is not present, did you forget to add it?');
        }
    }
    componentWillLoad() {
        this.hasFullWidthBg = hasSlot(this.el, 'full-width-bg');
    }
    render() {
        const { heading, preHeading, breadcrumbs, imgAlt, imgSrc, hasFullWidthBg, hideImgOnMobile } = this;
        return (h(Host, { key: 'df0c63d107f0e47d46d706aff84d6fef7b869ca5', class: {
                'has-full-width-bg': hasFullWidthBg,
                'hide-img-on-mobile': !!hideImgOnMobile,
            } }, h("div", { key: '6073db59c9f5f89f7452df8daa562942eb198110' }, hasFullWidthBg ? (h("div", { class: "full-width-bg" }, h("slot", { name: "full-width-bg" }))) : null, h("div", { key: '37cbac69b69a21dde06f445524ee55e54abdf242', class: "container" }, h("div", { key: '94f3a65f89050bcb9c256c268c350607535154e0', class: "hero-container" }, h("div", { key: '40d0542bbc423131fcea8265f845d626aeb56c30', class: "hero-text" }, hasFullWidthBg ? h("div", { class: "hero-text-bg" }) : null, breadcrumbs ? h("go-breadcrumbs", { items: breadcrumbs }) : null, h("div", { key: 'aafef086a6555ec3c3936d853e4b4ce7ac61cade', class: "pre-heading text-size-2" }, preHeading), h("h1", { key: '014c347f22abf2f5c878e02df625c1b4b563002c', class: "text-display-2" }, heading), h("slot", { key: 'db3091ebbea24d61a01107d022d6a013a8a7ce8c' })), imgSrc && imgAlt ? (h("div", { class: "hero-image" }, h("img", { src: imgSrc, alt: imgAlt, class: "featured-img" }))) : null)))));
    }
    get el() { return this; }
    static get watchers() { return {
        "imgSrc": ["watchImgSrc"],
        "imgAlt": ["watchImgAlt"]
    }; }
    static get style() { return GoHeroStyle0; }
}, [4, "go-hero", {
        "heading": [1],
        "preHeading": [1, "pre-heading"],
        "breadcrumbs": [1],
        "imgSrc": [1, "img-src"],
        "imgAlt": [1, "img-alt"],
        "hideImgOnMobile": [4, "hide-img-on-mobile"]
    }, undefined, {
        "imgSrc": ["watchImgSrc"],
        "imgAlt": ["watchImgAlt"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-hero", "go-breadcrumbs", "go-icon", "go-nav-link"];
    components.forEach(tagName => { switch (tagName) {
        case "go-hero":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoHero);
            }
            break;
        case "go-breadcrumbs":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "go-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "go-nav-link":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}
defineCustomElement();

export { GoHero as G, defineCustomElement as d };
