// @go-ui/core/dist/components/go-nav-item2.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { p as parseJsonProp, b as hasSlot } from './helper.js';
import { r as renderIcon } from './nav-helpers.js';
import { d as defineCustomElement$4 } from './go-icon2.js';
import { d as defineCustomElement$3 } from './go-nav-link2.js';
import { d as defineCustomElement$2 } from './go-nav-submenu2.js';
import { d as defineCustomElement$1 } from './go-nav-submenu-trigger2.js';
import { u as uniqueId } from './uniqueId.js';

const GoNavItem = /*@__PURE__*/ proxyCustomElement(class GoNavItem extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.navEvent = createEvent(this, "navigate", 7);
        this.subMenuToggleEvent = createEvent(this, "submenutoggle", 7);
        this.item = undefined;
        this.parsedItem = undefined;
        this.hasSubmenuSlot = false;
        this.isOpen = false;
    }
    parseItemProp() {
        var _a, _b;
        this.parsedItem = parseJsonProp(this.item);
        if ((_b = (_a = this.parsedItem) === null || _a === void 0 ? void 0 : _a.children) === null || _b === void 0 ? void 0 : _b.length) {
            this.submenuId = uniqueId('go-nav-item-submenu-');
        }
    }
    componentWillLoad() {
        this.parseItemProp();
        this.hasSubmenuSlot = hasSlot(this.el, 'submenu');
    }
    handleSubmenuToggle(isOpen) {
        console.log('event triggered', isOpen);
        this.isOpen = !!isOpen;
    }
    renderSubMenu(parent) {
        var _a;
        if (!parent) {
            return;
        }
        // if submenu item has children, render the current item and its children
        if (((_a = parent.children) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            return (h("div", { class: "submenu-list-container" }, h("div", { class: "submenu-header" }, h("go-nav-link", { block: true, item: parent })), h("ul", { class: "submenu-list" }, parent.children.map((child) => (h("li", null, h("go-nav-link", { block: true, item: child })))))));
        }
        return (h("div", { class: "nav-item" }, h("go-nav-link", { block: true, item: parent })));
    }
    render() {
        var _a;
        const { parsedItem: item, submenuId } = this;
        let Tag = 'a';
        const hasChildren = ((_a = item === null || item === void 0 ? void 0 : item.children) === null || _a === void 0 ? void 0 : _a.length) > 0 || this.hasSubmenuSlot;
        if (item === null || item === void 0 ? void 0 : item.isCurrent) {
            Tag = 'span';
        }
        let attrs = null;
        if (Tag === 'a') {
            attrs = Object.assign({ url: item === null || item === void 0 ? void 0 : item.url, onClick: (event) => {
                    this.navEvent.emit({ event, item });
                } }, item === null || item === void 0 ? void 0 : item.linkAttrs);
        }
        return (h(Host, { key: '3e6d846e2c3a73d24ca290cdc1a7dc6192b34cc9', role: "listitem", class: { 'nav-item': true, 'current': item === null || item === void 0 ? void 0 : item.isCurrent } }, h("slot", { key: 'f4f0d6596c6fe1b3a77661aeba8c33c3060b2985' }, hasChildren ? ([
            h("go-nav-submenu-trigger", { controls: submenuId }, h("span", { class: "nav-item-label" }, renderIcon(item === null || item === void 0 ? void 0 : item.icon), h("span", null, item === null || item === void 0 ? void 0 : item.label))),
            h("go-nav-submenu", { id: submenuId, columns: item === null || item === void 0 ? void 0 : item.columns }, h("go-nav-link", { slot: "submenu-header", block: true, item: item, showArrow: true }), item.children.map((child) => this.renderSubMenu(child))),
        ]) : (h(Tag, Object.assign({ class: "nav-item-inner" }, attrs), h("span", { class: "nav-item-label" }, renderIcon(item === null || item === void 0 ? void 0 : item.icon), h("span", null, item === null || item === void 0 ? void 0 : item.label)))))));
    }
    get el() { return this; }
    static get watchers() { return {
        "item": ["parseItemProp"]
    }; }
}, [4, "go-nav-item", {
        "item": [1],
        "parsedItem": [32],
        "hasSubmenuSlot": [32],
        "isOpen": [32]
    }, undefined, {
        "item": ["parseItemProp"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-nav-item", "go-icon", "go-nav-link", "go-nav-submenu", "go-nav-submenu-trigger"];
    components.forEach(tagName => { switch (tagName) {
        case "go-nav-item":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoNavItem);
            }
            break;
        case "go-icon":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "go-nav-link":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "go-nav-submenu":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "go-nav-submenu-trigger":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}
defineCustomElement();

export { GoNavItem as G, defineCustomElement as d };
