// @go-ui/core/dist/components/go-overlay2.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { i as inheritAttributes } from './helper.js';
import { d as trapFocus } from './dom.js';

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

// Older browsers don't support event options, feature detect it.

// Adopted and modified solution from Bohdan Didukh (2017)
// https://stackoverflow.com/questions/41594997/ios-10-safari-prevent-scrolling-behind-a-fixed-overlay-and-maintain-scroll-posi

var hasPassiveEvents = false;
if (typeof window !== 'undefined') {
  var passiveTestOptions = {
    get passive() {
      hasPassiveEvents = true;
      return undefined;
    }
  };
  window.addEventListener('testPassive', null, passiveTestOptions);
  window.removeEventListener('testPassive', null, passiveTestOptions);
}

var isIosDevice = typeof window !== 'undefined' && window.navigator && window.navigator.platform && (/iP(ad|hone|od)/.test(window.navigator.platform) || window.navigator.platform === 'MacIntel' && window.navigator.maxTouchPoints > 1);


var locks = [];
var documentListenerAdded = false;
var initialClientY = -1;
var previousBodyOverflowSetting = void 0;
var previousBodyPosition = void 0;
var previousBodyPaddingRight = void 0;

// returns true if `el` should be allowed to receive touchmove events.
var allowTouchMove = function allowTouchMove(el) {
  return locks.some(function (lock) {
    if (lock.options.allowTouchMove && lock.options.allowTouchMove(el)) {
      return true;
    }

    return false;
  });
};

var preventDefault = function preventDefault(rawEvent) {
  var e = rawEvent || window.event;

  // For the case whereby consumers adds a touchmove event listener to document.
  // Recall that we do document.addEventListener('touchmove', preventDefault, { passive: false })
  // in disableBodyScroll - so if we provide this opportunity to allowTouchMove, then
  // the touchmove event on document will break.
  if (allowTouchMove(e.target)) {
    return true;
  }

  // Do not prevent if the event has more than one touch (usually meaning this is a multi touch gesture like pinch to zoom).
  if (e.touches.length > 1) return true;

  if (e.preventDefault) e.preventDefault();

  return false;
};

var setOverflowHidden = function setOverflowHidden(options) {
  // If previousBodyPaddingRight is already set, don't set it again.
  if (previousBodyPaddingRight === undefined) {
    var _reserveScrollBarGap = !!options && options.reserveScrollBarGap === true;
    var scrollBarGap = window.innerWidth - document.documentElement.clientWidth;

    if (_reserveScrollBarGap && scrollBarGap > 0) {
      var computedBodyPaddingRight = parseInt(window.getComputedStyle(document.body).getPropertyValue('padding-right'), 10);
      previousBodyPaddingRight = document.body.style.paddingRight;
      document.body.style.paddingRight = computedBodyPaddingRight + scrollBarGap + 'px';
    }
  }

  // If previousBodyOverflowSetting is already set, don't set it again.
  if (previousBodyOverflowSetting === undefined) {
    previousBodyOverflowSetting = document.body.style.overflow;
    document.body.style.overflow = 'hidden';
  }
};

var restoreOverflowSetting = function restoreOverflowSetting() {
  if (previousBodyPaddingRight !== undefined) {
    document.body.style.paddingRight = previousBodyPaddingRight;

    // Restore previousBodyPaddingRight to undefined so setOverflowHidden knows it
    // can be set again.
    previousBodyPaddingRight = undefined;
  }

  if (previousBodyOverflowSetting !== undefined) {
    document.body.style.overflow = previousBodyOverflowSetting;

    // Restore previousBodyOverflowSetting to undefined
    // so setOverflowHidden knows it can be set again.
    previousBodyOverflowSetting = undefined;
  }
};

var setPositionFixed = function setPositionFixed() {
  return window.requestAnimationFrame(function () {
    // If previousBodyPosition is already set, don't set it again.
    if (previousBodyPosition === undefined) {
      previousBodyPosition = {
        position: document.body.style.position,
        top: document.body.style.top,
        left: document.body.style.left
      };

      // Update the dom inside an animation frame 
      var _window = window,
          scrollY = _window.scrollY,
          scrollX = _window.scrollX,
          innerHeight = _window.innerHeight;

      document.body.style.position = 'fixed';
      document.body.style.top = -scrollY;
      document.body.style.left = -scrollX;

      setTimeout(function () {
        return window.requestAnimationFrame(function () {
          // Attempt to check if the bottom bar appeared due to the position change
          var bottomBarHeight = innerHeight - window.innerHeight;
          if (bottomBarHeight && scrollY >= innerHeight) {
            // Move the content further up so that the bottom bar doesn't hide it
            document.body.style.top = -(scrollY + bottomBarHeight);
          }
        });
      }, 300);
    }
  });
};

var restorePositionSetting = function restorePositionSetting() {
  if (previousBodyPosition !== undefined) {
    // Convert the position from "px" to Int
    var y = -parseInt(document.body.style.top, 10);
    var x = -parseInt(document.body.style.left, 10);

    // Restore styles
    document.body.style.position = previousBodyPosition.position;
    document.body.style.top = previousBodyPosition.top;
    document.body.style.left = previousBodyPosition.left;

    // Restore scroll
    window.scrollTo(x, y);

    previousBodyPosition = undefined;
  }
};

// https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollHeight#Problems_and_solutions
var isTargetElementTotallyScrolled = function isTargetElementTotallyScrolled(targetElement) {
  return targetElement ? targetElement.scrollHeight - targetElement.scrollTop <= targetElement.clientHeight : false;
};

var handleScroll = function handleScroll(event, targetElement) {
  var clientY = event.targetTouches[0].clientY - initialClientY;

  if (allowTouchMove(event.target)) {
    return false;
  }

  if (targetElement && targetElement.scrollTop === 0 && clientY > 0) {
    // element is at the top of its scroll.
    return preventDefault(event);
  }

  if (isTargetElementTotallyScrolled(targetElement) && clientY < 0) {
    // element is at the bottom of its scroll.
    return preventDefault(event);
  }

  event.stopPropagation();
  return true;
};

var disableBodyScroll = function disableBodyScroll(targetElement, options) {
  // targetElement must be provided
  if (!targetElement) {
    // eslint-disable-next-line no-console
    console.error('disableBodyScroll unsuccessful - targetElement must be provided when calling disableBodyScroll on IOS devices.');
    return;
  }

  // disableBodyScroll must not have been called on this targetElement before
  if (locks.some(function (lock) {
    return lock.targetElement === targetElement;
  })) {
    return;
  }

  var lock = {
    targetElement: targetElement,
    options: options || {}
  };

  locks = [].concat(_toConsumableArray(locks), [lock]);

  if (isIosDevice) {
    setPositionFixed();
  } else {
    setOverflowHidden(options);
  }

  if (isIosDevice) {
    targetElement.ontouchstart = function (event) {
      if (event.targetTouches.length === 1) {
        // detect single touch.
        initialClientY = event.targetTouches[0].clientY;
      }
    };
    targetElement.ontouchmove = function (event) {
      if (event.targetTouches.length === 1) {
        // detect single touch.
        handleScroll(event, targetElement);
      }
    };

    if (!documentListenerAdded) {
      document.addEventListener('touchmove', preventDefault, hasPassiveEvents ? { passive: false } : undefined);
      documentListenerAdded = true;
    }
  }
};

var enableBodyScroll = function enableBodyScroll(targetElement) {
  if (!targetElement) {
    // eslint-disable-next-line no-console
    console.error('enableBodyScroll unsuccessful - targetElement must be provided when calling enableBodyScroll on IOS devices.');
    return;
  }

  locks = locks.filter(function (lock) {
    return lock.targetElement !== targetElement;
  });

  if (isIosDevice) {
    targetElement.ontouchstart = null;
    targetElement.ontouchmove = null;

    if (documentListenerAdded && locks.length === 0) {
      document.removeEventListener('touchmove', preventDefault, hasPassiveEvents ? { passive: false } : undefined);
      documentListenerAdded = false;
    }
  }

  if (isIosDevice) {
    restorePositionSetting();
  } else {
    restoreOverflowSetting();
  }
};

const goOverlayCss = "go-overlay{--go-overlay-color:rgba(var(--go-token-darkest), 0.8);--go-overlay-close-btn-text-color:var(--go-color-lightest);transition:opacity var(--go-duration-normal) var(--go-timing-function);position:fixed;top:0;left:0;right:0;bottom:0;width:0;height:0;overflow:auto;opacity:0;z-index:var(--layer-important)}@media (prefers-reduced-motion: reduce){go-overlay{transition-duration:0s}}go-overlay .close-btn-wrapper go-button{--go-btn-text-color:var(--go-overlay-close-btn-text-color)}go-overlay .overlay-bg{position:absolute;top:0;left:0;right:0;bottom:0}go-overlay .overlay-content{position:relative;visibility:hidden}go-overlay.active{background:var(--go-overlay-color);width:100%;height:100%;opacity:1}go-overlay.active .overlay-content{visibility:visible}:where([color-scheme=dark],[data-theme=dark]) go-overlay{--go-overlay-color:rgba(var(--go-token-darkest), 0.8);--go-overlay-close-btn-text-color:var(--go-text-color)}";
const GoOverlayStyle0 = goOverlayCss;

const GoOverlay = /*@__PURE__*/ proxyCustomElement(class GoOverlay extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.openEvent = createEvent(this, "overlayOpen", 7);
        this.closeEvent = createEvent(this, "overlayClose", 7);
        // Store attributes inherited from the host element
        this.inheritedAttrs = {};
        this.active = false;
        this.persistent = false;
    }
    componentWillLoad() {
        this.inheritedAttrs = inheritAttributes(this.el, ['class', 'style'], false);
        // move this.el to the end of the body
        const body = document.querySelector('body');
        body.appendChild(this.el);
    }
    componentDidLoad() {
        // close overlay on click outside
        if (!this.persistent) {
            this.el.addEventListener('click', (e) => {
                if (this.active) {
                    if (e.target === this.bgEl) {
                        e.preventDefault();
                        this.close();
                    }
                }
            });
        }
    }
    async close() {
        var _a;
        this.closeEvent.emit();
        enableBodyScroll(this.el);
        this.active = false;
        (_a = this.originator) === null || _a === void 0 ? void 0 : _a.focus();
        this.el.style.visibility = 'hidden';
    }
    async open() {
        this.originator = document.activeElement;
        this.openEvent.emit();
        disableBodyScroll(this.el);
        this.el.style.visibility = 'visible';
        this.active = true;
        // close overlay on escape
        if (!this.persistent) {
            document.addEventListener('keydown', (e) => {
                if (this.active) {
                    if (e.code === 'Escape') {
                        this.close();
                    }
                }
            });
        }
        trapFocus(this.el);
    }
    watchActiveHandler(newValue) {
        if (newValue) {
            this.open();
            return;
        }
        this.close();
    }
    render() {
        const { active, inheritedAttrs } = this;
        return (h(Host, Object.assign({ key: '40487b7dc4613247374351d9717a839b57134fee', class: { active } }, inheritedAttrs), h("div", { key: 'd987d6c1d03f40bde84eb192f45c994cf0ff0231', class: "overlay-bg", ref: (el) => (this.bgEl = el) }), h("div", { key: '101a5b3e63036d99086a31a44e8ab2b86282839d', class: "overlay-content" }, h("slot", { key: 'ca4041cc9384f00650321c3364e69b30bb946e95' }))));
    }
    get el() { return this; }
    static get watchers() { return {
        "active": ["watchActiveHandler"]
    }; }
    static get style() { return GoOverlayStyle0; }
}, [4, "go-overlay", {
        "active": [1540],
        "persistent": [4],
        "close": [64],
        "open": [64]
    }, undefined, {
        "active": ["watchActiveHandler"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-overlay"];
    components.forEach(tagName => { switch (tagName) {
        case "go-overlay":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoOverlay);
            }
            break;
    } });
}
defineCustomElement();

export { GoOverlay as G, defineCustomElement as d };
