// @go-ui/core/dist/components/go-skip-link.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { w as warning } from './helper.js';
import { u as uniqueId } from './uniqueId.js';

const goSkipLinkCss = "go-skip-link{--skip-link-padding:1rem 2rem;--skip-link-bg-color:var(--go-color-secondary-700);--skip-link-text-color:var(--go-color-lightest);position:fixed;top:1rem;left:1rem}go-skip-link a,go-skip-link a:focus-visible{display:block;padding:var(--skip-link-padding);background:var(--skip-link-bg-color);color:var(--skip-link-text-color);box-shadow:var(--shadow-2)}";
const GoSkipLinkStyle0 = goSkipLinkCss;

const GoSkipLink$1 = /*@__PURE__*/ proxyCustomElement(class GoSkipLink extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        /**
         * records if original target element has tabindex attribute set, if so, do not touch it
         */
        this.hasTabIndex = false;
        this.target = 'main';
        this.href = undefined;
    }
    componentWillLoad() {
        // get target element
        // check if id exists
        // if not, generate random id and set it in the link href
        const targetEl = document.querySelector(this.target);
        if (!targetEl) {
            warning(`[go-skip-link] target ${this.target} not found.`);
            return;
        }
        if (targetEl.hasAttribute('tabindex')) {
            this.hasTabIndex = true;
        }
        if (targetEl.id) {
            this.setTabIndex(targetEl, targetEl.id);
            return;
        }
        // generate a random id then set href
        const id = uniqueId('skip-link-target-');
        targetEl.id = id;
        this.setTabIndex(targetEl, id);
    }
    setTabIndex(targetEl, id) {
        this.href = `#${id}`;
        if (!this.hasTabIndex) {
            targetEl.setAttribute('tabindex', '-1');
        }
    }
    render() {
        const { href } = this;
        return (h(Host, { key: '142d6a31279bf07c74991b29663317f8478f5cce' }, h("a", { key: 'd76db3a9e306eb1297f7aed2ae14f7f39ca25fd9', href: href, class: "visually-hidden-focusable" }, h("slot", { key: '7ecbda922a7b31e3d0aea151f4acaed6b178ed8c' }))));
    }
    get el() { return this; }
    static get style() { return GoSkipLinkStyle0; }
}, [4, "go-skip-link", {
        "target": [1],
        "href": [32]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-skip-link"];
    components.forEach(tagName => { switch (tagName) {
        case "go-skip-link":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoSkipLink$1);
            }
            break;
    } });
}
defineCustomElement$1();

const GoSkipLink = GoSkipLink$1;
const defineCustomElement = defineCustomElement$1;

export { GoSkipLink, defineCustomElement };
