// @go-ui/core/dist/components/go-to-top.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { d as defineCustomElement$4 } from './go-button2.js';
import { d as defineCustomElement$3 } from './go-spinner2.js';
import { d as defineCustomElement$2 } from './go-tooltip2.js';
import { u as uniqueId } from './uniqueId.js';
import { d as debounce } from './floating-ui.dom.esm.js';

const goToTopCss = "go-to-top{--go-to-top-right:2rem;--go-to-top-bottom:2rem;display:block;position:fixed;bottom:var(--go-to-top-bottom);right:var(--go-to-top-right);pointer-events:none;z-index:var(--layer-important);opacity:0;transform:scale(0);transition:transform var(--go-duration-normal) var(--go-timing-function), opacity var(--go-duration-normal) var(--go-timing-function)}@media (prefers-reduced-motion: reduce){go-to-top{transition-duration:0s}}go-to-top.active{pointer-events:inherit;transform:scale(1);opacity:1}";
const GoToTopStyle0 = goToTopCss;

const GoToTop$1 = /*@__PURE__*/ proxyCustomElement(class GoToTop extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.offset = 200;
        this.active = false;
    }
    componentWillLoad() {
        // get body id or generate a new one if it doesn't exist
        this.targetId = document.body.getAttribute('id');
        if (!this.targetId) {
            this.targetId = uniqueId('body-');
            document.body.setAttribute('id', this.targetId);
        }
    }
    componentDidLoad() {
        window.onscroll = debounce(() => {
            this.active = window.scrollY > this.offset;
        }, 200);
    }
    render() {
        const { active } = this;
        return (h(Host, { key: '7477a98588e745f879bdbf469021d6bd3af8d6ae', class: { active } }, h("go-button", { key: '2ed2579eca645d6ca74471abc88d5eaea5481640', variant: "primary", flat: true, href: `#${this.targetId}`, id: "go-to-top-btn", type: "button", icon: true, round: true, "aria-labelledby": "go-to-top-btn-tooltip" }, h("svg", { key: '73cab3e976da71b09400d5750c491f199ac5ced7', xmlns: "http://www.w3.org/2000/svg", fill: "none", stroke: "currentColor", "stroke-linecap": "round", "stroke-linejoin": "round", "stroke-width": "2", viewBox: "0 0 24 24" }, h("path", { key: 'd53faba8f1b3fcd46ce2241e9d8cd82513351359', d: "m18 15-6-6-6 6" }))), h("go-tooltip", { key: 'd9b166056a1d6102a9a53a8d7641be33613e580f', id: "go-to-top-btn-tooltip", placement: "left", "trigger-id": "go-to-top-btn" }, "Back to top")));
    }
    get el() { return this; }
    static get style() { return GoToTopStyle0; }
}, [0, "go-to-top", {
        "offset": [2],
        "active": [32]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["go-to-top", "go-button", "go-spinner", "go-tooltip"];
    components.forEach(tagName => { switch (tagName) {
        case "go-to-top":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, GoToTop$1);
            }
            break;
        case "go-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "go-spinner":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "go-tooltip":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}
defineCustomElement$1();

const GoToTop = GoToTop$1;
const defineCustomElement = defineCustomElement$1;

export { GoToTop, defineCustomElement };
