// @go-ui/core/dist/collection/components/go-table-wrapper/go-table-wrapper.js
import { Host, h } from "@stencil/core";
export class GoTableWrapper {
    constructor() {
        this.striped = false;
        this.bordered = false;
        this.hoverable = false;
    }
    // Store attributes inherited from the host element
    componentWillLoad() { }
    render() {
        const { striped, bordered, hoverable } = this;
        return (h(Host, { key: 'c707d76b8c23c8d33cf1d059274ef2085a1a0568', class: {
                striped,
                bordered,
                hoverable,
            } }, h("slot", { key: '0e3b403c031b5ed09305ce96c98d4d5ebbcfc1b8' })));
    }
    static get is() { return "go-table-wrapper"; }
    static get originalStyleUrls() {
        return {
            "$": ["go-table-wrapper.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["go-table-wrapper.css"]
        };
    }
    static get properties() {
        return {
            "striped": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Striped rows"
                },
                "attribute": "striped",
                "reflect": false,
                "defaultValue": "false"
            },
            "bordered": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Bordered table"
                },
                "attribute": "bordered",
                "reflect": false,
                "defaultValue": "false"
            },
            "hoverable": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Hoverable rows"
                },
                "attribute": "hoverable",
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get elementRef() { return "el"; }
}
