// @go-ui/core/dist/collection/patterns/go-header-bar/go-header-bar.js
import { Host, h } from "@stencil/core";
export class GoHeaderBar {
    constructor() {
        this.breakpoint = 'desktop';
    }
    render() {
        const { breakpoint } = this;
        return (h(Host, { key: 'b3ac382116f2e788efac45b231828ee81da06955' }, h("div", { key: '8a4538cce87a81b9cfb5f333122824b76f913420', class: "container" }, h("div", { key: 'f9cdd80072f1e1ec475952eb842f366f8e8ba651', class: { 'header-bar': true, [`responsive-${breakpoint}`]: true } }, h("div", { key: '84e27006a45789631f33620cbed4f4a0d14ad416', class: {
                'mobile-menu-trigger': true,
                [`d-none-${breakpoint}`]: true,
            } }, h("slot", { key: 'ddde4bb48203fe2381e12ece60e29c4e5e97de47', name: "mobile-menu-trigger" })), h("div", { key: 'a559456a8ee4719b9bbadbdbae677a71be6cd5b2', class: "logo" }, h("slot", { key: '19936dc0ed47f77070000637412d0c4a8ae9e214', name: "logo" })), h("div", { key: 'd79db62f2714cfd4a7fb5f408b1c29c24a5c8824', class: "actions" }, h("slot", { key: '4c4ce9223fe59a1783f94d63822d93b663347435', name: "actions" })))), h("div", { key: '6b3332e4ed070c948797a581ac4c83e6c51aa7d5', class: {
                'nav-bar d-none': true,
                [`d-block-${breakpoint}`]: true,
            } }, h("slot", { key: '3a2e8ce040fd4d4e1c45b8c81e7a08b70fdef579', name: "nav-bar" }))));
    }
    static get is() { return "go-header-bar"; }
    static get originalStyleUrls() {
        return {
            "$": ["go-header-bar.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["go-header-bar.css"]
        };
    }
    static get properties() {
        return {
            "breakpoint": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "Breakpoints",
                    "resolved": "\"all\" | \"desktop\" | \"large\" | \"mobile\" | \"tablet\"",
                    "references": {
                        "Breakpoints": {
                            "location": "import",
                            "path": "../../interfaces",
                            "id": "src/interfaces/index.ts::Breakpoints"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls at which breakpoint the mobile menu (go-nav-drawer) becomes main nav menu (go-nav-bar)"
                },
                "attribute": "breakpoint",
                "reflect": false,
                "defaultValue": "'desktop'"
            }
        };
    }
    static get elementRef() { return "el"; }
}
