// @go-ui/core/dist/collection/components/go-search-bar/go-search-bar.js
import { Host, h } from "@stencil/core";
export class GoSearchBar {
    constructor() {
        this.isSearchFormOpen = false;
    }
    async openSearchForm() {
        this.isSearchFormOpen = true;
        this.el.focus();
    }
    async closeSearchForm() {
        this.isSearchFormOpen = false;
    }
    render() {
        const { isSearchFormOpen: open } = this;
        return (h(Host, { key: '84a6831ad5d788acda9d09b091f133747ea6a7de', class: {
                open,
            } }, h("go-overlay", { key: 'de14ff282e21f9d994ca526573a1d0ccf090e58e', active: open ? true : false, class: "go-search-bar-overlay", onOverlayClose: () => this.closeSearchForm() }, h("slot", { key: '03f373fff8ce617042a8da4d67be2baae3be8067', name: "search-form" }), h("div", { key: '82a69c0b18ac00c3303c73f217c34ccc86bbace5', class: "close-btn-wrapper" }, h("go-button", { key: '7204562ee9b5f547d312ec8ad155e93ee54e07ba', flat: true, stack: true, variant: "text", compact: true, onClick: () => this.closeSearchForm() }, h("svg", { key: '49074ba796b0211d6fcbb36c364d862b6a5027c1', "aria-hidden": "true", xmlns: "http://www.w3.org/2000/svg", fill: "none", stroke: "currentColor", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round", viewBox: "0 0 24 24" }, h("path", { key: '5cd9880916200a8250ea6dd74435fb9cfe0eacc5', d: "M18 6 6 18M6 6l12 12" })), h("span", { key: '8d092da2dc9f627d58d998027b21e65660dcf192' }, "Close")))), h("div", { key: 'd2aad50d5556bf17c9d9f485f57ee2040bd1c7ae', class: "search-btn" }, h("go-button", { key: '00e15f35f6a7e9299108b28e09daaa15427fd924', class: "open-btn", compact: true, flat: true, stack: true, variant: "text", onClick: () => this.openSearchForm() }, h("svg", { key: 'da2273a949dad89b3bd5d380c5d9be4b6212403c', xmlns: "http://www.w3.org/2000/svg", viewBox: "0 0 24 24", fill: "currentColor" }, h("path", { key: '01607ec5b73ee49a3bf78e1ef6e34fc06eb1a93f', d: "M0 0h24v24H0z", fill: "none" }), h("path", { key: '296619c0d6ad69128687a682acba3db303e476b5', d: "M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z" })), h("span", { key: 'fd7eefc643f7277102766cf97c97f0956feaec8e' }, "Search")))));
    }
    static get is() { return "go-search-bar"; }
    static get originalStyleUrls() {
        return {
            "$": ["go-search-bar.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["go-search-bar.css"]
        };
    }
    static get states() {
        return {
            "isSearchFormOpen": {}
        };
    }
    static get methods() {
        return {
            "openSearchForm": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "",
                    "tags": []
                }
            },
            "closeSearchForm": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "el"; }
}
