// @go-ui/core/dist/esm/go-footer.entry.js
import { r as registerInstance, h, H as Host, g as getElement } from './index-a73e6b51.js';
import { p as parseJsonProp, w as warning, b as hasSlot } from './helper-bc7b4586.js';
import './index-30633753.js';
import './uniqueId-0b35bc55.js';

const goFooterCss = "go-footer{--footer-nav-section-padding:3rem 0;--footer-bottom-section-padding:2rem 0;--footer-copyright-section-padding:2rem 0;--footer-bg-color:var(--go-color-lightest);--footer-text-color:var(--go-color-text);--footer-separator-color:var(--go-color-neutral-300);display:block;background:var(--footer-bg-color);color:var(--footer-text-color)}go-footer go-nav-link{--nav-link-padding:0.5em}go-footer a,go-footer go-nav-list{--go-color-link:var(--footer-text-color);--nav-list-header-color:var(--footer-text-color)}go-footer a .nav-link-text-description,go-footer go-nav-list .nav-link-text-description{--nav-link-description-color:var(--footer-text-color)}go-footer .nav-container{padding:var(--footer-nav-section-padding)}go-footer .footer-bottom{border-top:0.0625rem solid var(--footer-separator-color);padding:var(--footer-bottom-section-padding)}go-footer .copyright{border-top:0.0625rem solid var(--footer-separator-color);padding:var(--footer-copyright-section-padding)}";
const GoFooterStyle0 = goFooterCss;

const GoFooter = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.hasCopyRightSlot = false;
        this.hasFooterBottomSlot = false;
        this.links = undefined;
        this.navLabel = 'Footer';
        this.navCols = 1;
        this.navColsTablet = 3;
        this.navColsDesktop = 4;
        this.listHeadingTag = 'h4';
        this.navItems = undefined;
    }
    async watchItems(newItems) {
        this.navItems = parseJsonProp(newItems);
    }
    componentWillLoad() {
        var _a;
        this.navItems = parseJsonProp(this.links);
        if (((_a = this.navItems) === null || _a === void 0 ? void 0 : _a.length) > 0 && !this.navLabel) {
            warning('Please add a unique "nav-label" in order to put navigation items into the nav landmark for better accessibility.');
        }
        // check if copyright slot is empty
        this.hasCopyRightSlot = hasSlot(this.el, 'copyright');
        this.hasFooterBottomSlot = hasSlot(this.el, 'footer-bottom');
    }
    render() {
        const { navItems, navLabel, hasCopyRightSlot, hasFooterBottomSlot, navCols, navColsTablet, navColsDesktop } = this;
        return (h(Host, { key: '80b249ff0ad0f4968770ef8d5831ddb689a03e0c' }, h("footer", { key: 'bc860de5eec9caf5554a7a0994b7d6818c8fff7b' }, navItems ? (h("div", { class: "container nav-container" }, h("nav", { "aria-label": navLabel }, h("div", { class: "row" }, navItems === null || navItems === void 0 ? void 0 : navItems.map((item) => (h("go-nav-list", { block: true, class: `col-${12 / navCols} col-tablet-${12 / navColsTablet} col-desktop-${12 / navColsDesktop}`, items: item === null || item === void 0 ? void 0 : item.children }, h("go-nav-link", { block: true, slot: "header", item: item, showArrow: true })))))))) : (h("slot", null)), hasFooterBottomSlot ? (h("div", { class: "container" }, h("div", { class: "footer-bottom" }, h("slot", { name: "footer-bottom" })))) : null, hasCopyRightSlot ? (h("div", { class: "container" }, h("div", { class: "copyright" }, h("slot", { name: "copyright" })))) : null)));
    }
    get el() { return getElement(this); }
    static get watchers() { return {
        "links": ["watchItems"]
    }; }
};
GoFooter.style = GoFooterStyle0;

export { GoFooter as go_footer };
