// @go-ui/core/dist/esm/go-table-wrapper.entry.js
import { r as registerInstance, h, H as Host, g as getElement } from './index-a73e6b51.js';

const goTableWrapperCss = "go-table-wrapper{display:block;overflow:auto}go-table-wrapper.striped>table tr:nth-child(even){background-color:var(--go-color-table-striped-bg)}go-table-wrapper.hoverable>table tbody tr:hover,go-table-wrapper.hoverable>table tbody tr:focus{background-color:var(--go-color-table-hover-bg)}go-table-wrapper.bordered>table th,go-table-wrapper.bordered>table td,go-table-wrapper.bordered>table tr{border-width:var(--go-table-border-width-thin) 0;border-color:var(--go-color-table-border);border-style:solid}go-table-wrapper.bordered>table thead th{border-width:0 var(--go-table-border-width-thin)}go-table-wrapper.bordered>table tbody td,go-table-wrapper.bordered>table tbody th{border-width:0 var(--go-table-border-width-thin)}";
const GoTableWrapperStyle0 = goTableWrapperCss;

const GoTableWrapper = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.striped = false;
        this.bordered = false;
        this.hoverable = false;
    }
    // Store attributes inherited from the host element
    componentWillLoad() { }
    render() {
        const { striped, bordered, hoverable } = this;
        return (h(Host, { key: 'c707d76b8c23c8d33cf1d059274ef2085a1a0568', class: {
                striped,
                bordered,
                hoverable,
            } }, h("slot", { key: '0e3b403c031b5ed09305ce96c98d4d5ebbcfc1b8' })));
    }
    get el() { return getElement(this); }
};
GoTableWrapper.style = GoTableWrapperStyle0;

export { GoTableWrapper as go_table_wrapper };
