// @go-ui/core/dist/esm/go-tab.entry.js
import { r as registerInstance, h, H as Host, g as getElement } from './index-a73e6b51.js';

const goTabCss = "go-tab{display:none}go-tab.active{display:block}go-tab:focus-visible{outline-offset:0.5rem}";
const GoTabStyle0 = goTabCss;

const GoTab = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.label = undefined;
        this.active = undefined;
        this.tabId = undefined;
        this.panelId = undefined;
        this.iconPosition = 'before';
        this.iconOnly = false;
    }
    async setActive(active) {
        this.active = active;
    }
    render() {
        const { panelId, tabId, active } = this;
        return (h(Host, { key: '260e446d24b1228c702c45b1ee98094be0137e17', tabindex: "0", role: "tabpanel", id: panelId, "aria-labelledby": tabId, class: { 'tab-panel': true, active } }, h("slot", { key: '98cf8d2a43908aa5dc6371f4f337d5029240cef2' })));
    }
    get el() { return getElement(this); }
};
GoTab.style = GoTabStyle0;

export { GoTab as go_tab };
