// @go-ui/core/dist/cjs/form-6e13a9ec.js
'use strict';

const helper = require('./helper-c1931cd5.js');
const kebabCase = require('./kebabCase-f98d4407.js');

const fieldSlotNames = ['icon-before', 'icon-after', 'prefix', 'suffix', 'hint', 'label'];
const fieldPropNames = [
    'prefix',
    'value',
    'error',
    'controlId',
    'labelId',
    'prefixId',
    'suffixId',
    'hintId',
    'errorId',
    'name',
    'label',
    'disabled',
    'hint',
    'readonly',
];
function loadFieldProps(compInstance) {
    let props = {};
    fieldPropNames.forEach((propKey) => {
        props[propKey] = compInstance[propKey];
    });
    const specialPropMap = {
        // controlId: 'id',
        idPrefix: 'prefix',
    };
    Object.keys(specialPropMap).forEach((key) => {
        const instancePropKey = specialPropMap[key];
        props[key] = compInstance[instancePropKey];
    });
    return props;
}
function loadFieldSlots(rootEl) {
    let result = {};
    fieldSlotNames.forEach((slotName) => {
        result[slotName] = helper.hasSlot(rootEl, slotName);
    });
    return result;
}
function inheritNonFieldAttrs(compInstance, ignoreAttrs = []) {
    return helper.inheritComponentAttrs(compInstance, [...fieldPropNames.map(kebabCase.kebabCase), ...ignoreAttrs]);
}

exports.fieldPropNames = fieldPropNames;
exports.fieldSlotNames = fieldSlotNames;
exports.inheritNonFieldAttrs = inheritNonFieldAttrs;
exports.loadFieldProps = loadFieldProps;
exports.loadFieldSlots = loadFieldSlots;
