// @go-ui/core/dist/cjs/go-accordion.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-348d3f4f.js');
const helper = require('./helper-c1931cd5.js');
require('./index-c48c451e.js');
require('./uniqueId-93e4ecef.js');

const goAccordionCss = "go-accordion{--heading-padding-x:1rem;--heading-padding-y:var(--go-size--1);--panel-padding-x:1rem;--panel-padding-y:1rem;--panel-max-height:100vh;--heading-text-color:var(--go-color-text, #333);--heading-bg-color:var(--go-color-neutral-100);--heading-font-size:var(--go-size-0, 1rem);--heading-font-weight:700;--panel-bg-color:var(--go-color-bg);--heading-border-bottom:var(--go-border-thin) solid var(--go-color-border);--heading-bg-color-hover:var(--go-color-neutral-200);--heading-fg-color-hover:var(--go-color-text);--heading-radius:var(--radius-2)}";
const GoAccordionStyle0 = goAccordionCss;

const GoAccordion = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.multiple = false;
    }
    componentWillLoad() {
        this.items = Array.from(this.el.children);
    }
    componentDidLoad() {
        // if there's any active accordion item, show them by default and close the rest.
        if (!this.multiple) {
            this.activeItem = helper.selectDirectChildren(this.el, 'go-accordion-item[active]')[0];
            if (this.activeItem) {
                this.activeItem.open();
            }
            this.closeNonActive();
        }
    }
    openHandler(e) {
        e.stopPropagation();
        const eventTarget = e.target;
        if (!this.multiple) {
            this.activeItem = eventTarget;
            this.closeNonActive();
        }
    }
    closeNonActive() {
        this.items.forEach((item) => {
            if (!item.isSameNode(this.activeItem)) {
                item.close();
            }
        });
    }
    render() {
        return (index.h(index.Host, { key: '3ac7b78f920fe5f4ec3e9c9374a2fe9eb554d168' }, index.h("slot", { key: '3eeee3015929d1e851905a0f4ef6d75cb726191b' })));
    }
    get el() { return index.getElement(this); }
};
GoAccordion.style = GoAccordionStyle0;

exports.go_accordion = GoAccordion;
