// @go-ui/core/dist/cjs/go-breadcrumbs.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-348d3f4f.js');
const helper = require('./helper-c1931cd5.js');
require('./index-c48c451e.js');
require('./uniqueId-93e4ecef.js');

const goBreadcrumbsCss = "go-breadcrumbs{--breadcrumb-item-gap:0.5rem;display:block}go-breadcrumbs go-link{--nav-link-current-bar-width:0}go-breadcrumbs ol{padding:0;margin:0;list-style:none}go-breadcrumbs ol li{align-items:center;vertical-align:middle;display:none}@media (min-width: 600px){go-breadcrumbs ol li{display:inline-flex}}go-breadcrumbs ol li:last-child{display:inline-flex;margin-right:0}go-breadcrumbs ol li:nth-last-child(2){display:inline-flex}go-breadcrumbs ol li .breadcrumb-separator{margin-left:var(--breadcrumb-item-gap);margin-right:var(--breadcrumb-item-gap);width:1rem;height:1rem;vertical-align:middle}";
const GoBreadcrumbsStyle0 = goBreadcrumbsCss;

const GoBreadcrumb = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.label = 'Breadcrumbs';
        this.items = undefined;
        this.hideCurrent = false;
        this.navItems = [];
    }
    async watchItems(newItems) {
        this.navItems = this.getItems(newItems);
    }
    componentWillLoad() {
        this.navItems = this.getItems(this.items);
    }
    getItems(items) {
        const navItems = helper.parseJsonProp(items);
        if (!this.hideCurrent) {
            return navItems;
        }
        return navItems.filter((item, i) => item.url || i !== navItems.length - 1);
    }
    render() {
        const { label, navItems, hideCurrent } = this;
        if (!(navItems === null || navItems === void 0 ? void 0 : navItems.length)) {
            return null;
        }
        return (index.h(index.Host, null, index.h("nav", { "aria-label": label }, index.h("ol", null, navItems.map((item, i) => {
            const isLast = i === navItems.length - 1;
            const isCurrentPageALink = isLast && item.url && !hideCurrent;
            return (index.h("li", null, index.h("go-nav-link", { item: item, "aria-current": isCurrentPageALink ? 'page' : null }), i < navItems.length - 1 ? (index.h("svg", { class: "breadcrumb-separator", xmlns: "http://www.w3.org/2000/svg", fill: "none", stroke: "currentColor", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round", viewBox: "0 0 24 24" }, index.h("path", { d: "m9 18 6-6-6-6" }))) : null));
        })))));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "items": ["watchItems"]
    }; }
};
GoBreadcrumb.style = GoBreadcrumbsStyle0;

exports.go_breadcrumbs = GoBreadcrumb;
