// @go-ui/core/dist/cjs/go-blockquote.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-348d3f4f.js');

const goBlockquoteCss = "@charset \"UTF-8\";go-blockquote{--blockquote-padding:1rem 2rem;--blockquote-border:var(--border-size-3) solid var(--go-color-neutral-400);--blockquote-caption-font-size:var(--go-size-0);--blockquote-font-size:var(--go-size-1);--blockquote-icon-size:5rem;--blockquote-caption-margin-top:1rem;display:flex;position:relative;font-style:italic;border-left:var(--blockquote-border);align-items:center;padding:var(--blockquote-padding)}go-blockquote.quote-mark::before{content:\"❝\"/\"\";font-size:var(--blockquote-icon-size);line-height:1;position:absolute;top:0;left:0;z-index:-1;opacity:0.2}go-blockquote figure,go-blockquote blockquote{border:none;padding:0;margin:0}go-blockquote figcaption{margin-top:var(--blockquote-caption-margin-top);display:flex;gap:0.5em;font-size:var(--blockquote-caption-font-size)}go-blockquote blockquote{font-size:var(--blockquote-font-size)}";
const GoBlockquoteStyle0 = goBlockquoteCss;

const GoBlockquote = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.citeUrl = undefined;
        this.hideQuoteMark = false;
    }
    render() {
        return (index.h(index.Host, { key: '78a03dda92ba3ff2f77b9d8c4a0f720a1ceecd96', class: { 'quote-mark': !this.hideQuoteMark } }, index.h("figure", { key: 'cc032a20cb02a3bf56d19f145c41745531ef886b' }, index.h("blockquote", { key: '24218268c913c37345cdd8f05196178b05b5919c', cite: this.citeUrl }, index.h("slot", { key: '4469730492c1603f68e98bd113fd9c54d63a7498' })), index.h("figcaption", { key: 'e26d76df13d8920f4d5e1dc2944ab12a24044786' }, index.h("span", { key: '042166bd73f1f1cce8fa02dc0e5c864bada57fe1' }, "\u2014"), index.h("slot", { key: '074bf22e966f2d17b38c5eef9b27627e19f4e25a', name: "caption" })))));
    }
    get el() { return index.getElement(this); }
};
GoBlockquote.style = GoBlockquoteStyle0;

exports.go_blockquote = GoBlockquote;
