// @go-ui/core/dist/cjs/go-dropdown.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-348d3f4f.js');
const floatingUi_dom_esm = require('./floating-ui.dom.esm-bae85b62.js');
require('./index-c48c451e.js');
const dom = require('./dom-7cff1398.js');
const uniqueId = require('./uniqueId-93e4ecef.js');
const debounce = require('./debounce-7cb0ceca.js');

const goDropdownCss = "go-dropdown{--dropdown-shadow:var(--shadow-3);--dropdown-bg-color:var(--go-color-lightest);--dropdown-duration:var(--go-duration-fast);--dropdown-radius:var(--radius-2);position:fixed;display:none;visibility:hidden;opacity:0;box-shadow:var(--dropdown-shadow);pointer-events:none;z-index:1;transform:translateY(-0.3125rem);transform-origin:top;width:var(--dropdown-width, 200px);background-color:var(--dropdown-bg-color);border-radius:var(--dropdown-radius);transition:opacity var(--dropdown-duration) var(--go-timing-function), visibility var(--go-duration-normal) var(--go-timing-function), transform var(--dropdown-duration) var(--go-timing-function)}@media (prefers-reduced-motion: reduce){go-dropdown{transition-duration:0s}}go-dropdown.full-width{--dropdown-width:100%}go-dropdown.flip{transform-origin:bottom;transform:translateY(0.3125rem)}go-dropdown.is-active{visibility:visible;opacity:1;pointer-events:inherit;transform:translateY(0)}";
const GoDropdownStyle0 = goDropdownCss;

const GoDropdown = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.opened = index.createEvent(this, "opened", 7);
        this.closed = index.createEvent(this, "closed", 7);
        this.triggerSelector = undefined;
        this.isActive = false;
        this.width = 'auto';
        this.noTriggerClickHandler = false;
        this.fullWidth = false;
        this.disabled = false;
    }
    /**
     * opens dropdown
     */
    async open() {
        if (this.disabled) {
            return;
        }
        this.originEl = document.activeElement;
        this.isActive = true;
        this.opened.emit();
    }
    /**
     * closes dropdown
     */
    async close(focusBackToTrigger = false) {
        this.isActive = false;
        this.closed.emit();
        if (this.originEl && focusBackToTrigger) {
            this.originEl.focus();
        }
    }
    /**
     * toggles dropdown
     */
    async toggle() {
        if (this.isActive) {
            this.close();
        }
        else {
            this.open();
        }
    }
    componentWillLoad() {
        // if id attribute is not provided, generate a unique id for the dropdown
        if (!this.el.id) {
            this.el.id = uniqueId.uniqueId('go-dropdown-');
        }
        this.triggerEl = document.querySelector(`${this.triggerSelector}`);
    }
    async componentDidLoad() {
        if (!this.triggerEl) {
            return;
        }
        await this.init();
        // press esc to close dropdown
        this.escapeHandler = (e) => {
            if (e.code === 'Escape') {
                this.close();
            }
        };
        this.focusOutHandler = (e) => {
            if (this.triggerEl.contains(e.relatedTarget)) {
                return;
            }
            if (e.relatedTarget && !this.el.contains(e.relatedTarget)) {
                this.close();
            }
        };
        document.addEventListener('keydown', this.escapeHandler);
        // focus outside to close dropdown
        this.el.addEventListener('focusout', this.focusOutHandler);
    }
    disconnectedCallback() {
        if (this.escapeHandler) {
            document.removeEventListener('keydown', this.escapeHandler);
        }
        if (this.focusOutHandler) {
            this.el.removeEventListener('focusout', this.focusOutHandler);
        }
        if (this.clickOutsideCleanUp) {
            this.clickOutsideCleanUp();
        }
        if (this.cleanupAutoUpdate) {
            this.cleanupAutoUpdate();
        }
    }
    async init() {
        this.triggerEl.setAttribute('aria-haspopup', 'true');
        this.setTriggerExpanded(this.isActive);
        // add click event listener
        if (!this.noTriggerClickHandler) {
            this.triggerEl.addEventListener('click', () => this.toggle());
        }
        this.clickOutsideCleanUp = dom.onClickOutside(this.el, (e) => {
            if (!this.triggerEl.contains(e.target) && this.isActive) {
                this.close();
            }
        });
        /**
         * Calculate position of dropdown
         */
        const middleware = [floatingUi_dom_esm.offset(4), floatingUi_dom_esm.flip()];
        this.cleanupAutoUpdate = floatingUi_dom_esm.autoUpdate(this.triggerEl, this.el, debounce.debounce(() => {
            floatingUi_dom_esm.computePosition(this.triggerEl, this.el, {
                strategy: 'fixed',
                placement: 'bottom-start',
                middleware,
            }).then(({ x, y, middlewareData }) => {
                var _a, _b;
                Object.assign(this.el.style, {
                    left: `${x}px`,
                    top: `${y}px`,
                });
                if (((_b = (_a = middlewareData.flip) === null || _a === void 0 ? void 0 : _a.overflows) === null || _b === void 0 ? void 0 : _b.length) > 0) {
                    this.el.classList.add('flip');
                }
                else {
                    this.el.classList.remove('flip');
                }
            });
        }, 50));
    }
    setTriggerExpanded(expanded) {
        this.triggerEl.setAttribute('aria-expanded', String(expanded));
    }
    handleActiveChange(isActive) {
        this.setTriggerExpanded(isActive);
        if (isActive) {
            this.el.style.display = 'block';
            // focusFirstWithin(this.el);
        }
        else {
            this.el.addEventListener('transitionend', () => {
                this.el.style.display = 'none';
            }, { once: true });
            this.triggerEl.focus();
        }
    }
    render() {
        let { isActive, width, fullWidth } = this;
        return (index.h(index.Host, { key: '174d6d272f5936aec0445bd40168e6f929adc093', class: { 'is-active': isActive, 'full-width': fullWidth }, "aria-hidden": isActive ? 'false' : 'true', style: { '--dropdown-width': width } }, index.h("div", { key: '1fd09ef5f152240d046dafdb14ca90cf9d39479d', class: "dropdown-content" }, index.h("slot", { key: 'c5d8cbee34c6724319ee63bddb84fc8d075f97d4' }))));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "isActive": ["handleActiveChange"]
    }; }
};
GoDropdown.style = GoDropdownStyle0;

exports.go_dropdown = GoDropdown;
