// @go-ui/core/dist/cjs/go-heading-row.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-348d3f4f.js');

const goHeadingRowCss = "go-heading-row{--heading-row-alignment:center;--heading-row-justify:space-between;--heading-row-gap-bottom:var(--go-size-1);--heading-row-inner-gap:0 var(--go-outer-gutter);--heading-row-inner-gap-wide:var(--heading-row-inner-gap)}go-heading-row .heading-row{display:block;margin-bottom:var(--heading-row-gap-bottom);padding:var(--heading-row-inner-gap)}@media (min-width: 600px){go-heading-row .heading-row{display:flex;align-items:var(--heading-row-alignment);justify-content:var(--heading-row-justify);padding:var(--heading-row-inner-gap-wide)}}";
const GoHeadingRowStyle0 = goHeadingRowCss;

const GoHeadingRow = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.heading = undefined;
        this.headingTag = 'h2';
        this.moreLinkHref = undefined;
        this.moreLinkText = undefined;
    }
    componentWillLoad() {
        this.hasHeadingSlot = !!this.el.querySelector('slot[name="heading"]');
        this.hasActionSlot = !!this.el.querySelector('slot[name="action"]');
    }
    render() {
        const { heading, moreLinkHref, moreLinkText, hasHeadingSlot, hasActionSlot, headingTag: HeadingTag } = this;
        const shouldRenderHeading = heading || hasHeadingSlot;
        const shouldRenderMoreLink = (moreLinkHref && moreLinkText) || hasActionSlot;
        return (index.h(index.Host, { key: '5e2070f28506c8f26bb35f5b7aacb7b4680dea05' }, shouldRenderHeading || shouldRenderMoreLink ? (index.h("div", { class: "heading-row" }, shouldRenderHeading ? (index.h("slot", { name: "heading" }, index.h(HeadingTag, null, heading))) : (index.h("span", null)), shouldRenderMoreLink ? (index.h("slot", { name: "action" }, index.h("go-nav-link", { item: {
                url: moreLinkHref,
                label: moreLinkText,
            }, showArrow: true }))) : null)) : null));
    }
    get el() { return index.getElement(this); }
};
GoHeadingRow.style = GoHeadingRowStyle0;

exports.go_heading_row = GoHeadingRow;
