// @go-ui/core/dist/cjs/go-link.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-348d3f4f.js');
const helper = require('./helper-c1931cd5.js');
require('./index-c48c451e.js');
require('./uniqueId-93e4ecef.js');

const goLinkCss = "go-link .external-link-icon{height:0.9em;fill:currentColor;transform:translate(10%, 15%);margin-left:0.25em}go-link .expand::after{content:\"\";position:absolute;top:0;right:0;bottom:0;left:0}";
const GoLinkStyle0 = goLinkCss;

const GoLink = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        // Store attributes inherited from the host element
        this.attrs = {};
        this.href = undefined;
        this.target = undefined;
        this.expandClickableArea = false;
        this.isExternal = false;
        this.isNewTab = false;
    }
    componentWillLoad() {
        this.checkExternal();
        this.attrs = helper.inheritAttributes(this.el, ['href', 'target', 'expand-clickable-area']);
    }
    checkNewTab() {
        this.isNewTab = this.target === '_blank';
    }
    checkExternal() {
        if (!this.href) {
            return;
        }
        this.isExternal = this.isExternalURL(this.href);
        if (this.isExternal) {
            this.isNewTab = true;
        }
    }
    isExternalURL(url) {
        if (url.startsWith('#')) {
            return false;
        }
        const tmp = document.createElement('a');
        tmp.href = url;
        const isExternal = tmp.host !== window.location.host;
        tmp.remove();
        return isExternal;
    }
    render() {
        const { href, target, attrs, isExternal, isNewTab, expandClickableArea } = this;
        let classes = ['go-link'];
        if (attrs['class']) {
            classes.push(attrs['class']);
        }
        if (expandClickableArea) {
            classes.push('expand');
        }
        let linkAttrs = Object.assign(Object.assign({}, attrs), { href, class: classes.join(' '), target: isExternal || isNewTab ? '_blank' : target });
        return (index.h("a", Object.assign({ key: '518827946136783415ce3df61775299b55c0651c' }, linkAttrs), index.h("slot", { key: 'd4ced2692a1ab8d69646a6ab5396353b914ee15b' }), isExternal || isNewTab
            ? [
                index.h("svg", { xmlns: "http://www.w3.org/2000/svg", viewBox: "0 0 24 24", "aria-hidden": "true", class: "external-link-icon" }, index.h("path", { d: "M0 0h24v24H0z", fill: "none" }), index.h("path", { d: "M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z" })),
                index.h("span", { class: "visually-hidden" }, "Opens in new a tab or window"),
            ]
            : null));
    }
    get el() { return index.getElement(this); }
    static get watchers() { return {
        "target": ["checkNewTab"],
        "href": ["checkExternal"]
    }; }
};
GoLink.style = GoLinkStyle0;

exports.go_link = GoLink;
