// @go-ui/core/dist/cjs/go-nav-submenu_2.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-348d3f4f.js');
const helper = require('./helper-c1931cd5.js');
const dom = require('./dom-7cff1398.js');
require('./index-c48c451e.js');
require('./uniqueId-93e4ecef.js');

const GoNavSubmenu = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.toggleEvent = index.createEvent(this, "toggle", 7);
        this.clickOutsideCleanUp = null;
        this.escapeCleanUp = null;
        this.hasHeaderSlot = false;
        this.columns = 1;
        this.isOpen = false;
    }
    async open() {
        this.isOpen = true;
        this.toggleEvent.emit({ isOpen: true });
        window.requestAnimationFrame(() => {
            // click outside to close menus
            this.clickOutsideCleanUp = dom.onClickOutside(this.el, () => this.close());
            // esc to close menus
            this.escapeCleanUp = dom.onEscape(document, () => this.close());
        });
    }
    async close() {
        this.isOpen = false;
        this.toggleEvent.emit({ isOpen: false });
        this.clickOutsideCleanUp && this.clickOutsideCleanUp();
        this.escapeCleanUp && this.escapeCleanUp();
    }
    async toggle() {
        if (this.isOpen) {
            this.close();
        }
        else {
            this.open();
        }
    }
    componentWillLoad() {
        this.parentNavItem = this.el.closest('go-nav-item');
        if (!this.parentNavItem) {
            helper.warning('<go-nav-submenu> must be a child of <go-nav-item>', this.el);
            return;
        }
        this.hasHeaderSlot = helper.hasSlot(this.el, 'submenu-header');
    }
    render() {
        const { columns, isOpen, hasHeaderSlot } = this;
        return (index.h("div", { key: 'ac5bc09df1a4ebbb7996e60832e3a2899f38bfa1', class: { 'submenu-container': true, 'open': isOpen }, style: { '--submenu-columns': String(columns) } }, hasHeaderSlot ? (index.h("div", { class: "submenu-header" }, index.h("slot", { name: "submenu-header" }))) : null, index.h("div", { key: '196f1c3f4ebc2de3d75de0967655f1b6fd08d95c', class: "submenu-body" }, index.h("slot", { key: '00194d39b83582368dce9e42e9eb2e22c01d87f7' }))));
    }
    get el() { return index.getElement(this); }
};

const GoNavSubmenuTrigger = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.controls = undefined;
        this.isOpen = false;
    }
    loadSubmenuEl() {
        if (this.controls) {
            this.submenuEl = document.getElementById(this.controls);
            if (!this.submenuEl) {
                helper.warning('<go-nav-submenu-trigger> is missing <go-nav-submenu> with id: ' + this.controls, this.el);
            }
            this.submenuEl.addEventListener('toggle', (e) => {
                this.isOpen = e.detail.isOpen;
            });
        }
    }
    toggleOpenState() {
        this.submenuEl.toggle();
    }
    componentDidLoad() {
        this.loadSubmenuEl();
    }
    render() {
        const Tag = 'button';
        return (index.h(Tag, { key: '0afb1549dd2e97a8e0b65ab254720b894be26040', class: "nav-item-inner", onClick: () => this.toggleOpenState(), type: "button", "aria-haspopup": "true", "aria-controls": this.controls ? this.controls : undefined, "aria-expanded": this.isOpen ? 'true' : 'false' }, index.h("span", { key: '6bbbeafa1cdfccd4b2cf37440863a8f5154e3e4e', class: "nav-item-label" }, index.h("slot", { key: '685ccffaf7270fee2aa2106aa6c0e2bafe21b5e5' })), index.h("svg", { key: '9f8b1a268ad2f5a790c13fda3b6839a774e52f8b', xmlns: "http://www.w3.org/2000/svg", fill: "none", stroke: "currentColor", "stroke-width": "2", "stroke-linecap": "round", "stroke-linejoin": "round", viewBox: "0 0 24 24" }, index.h("path", { key: 'd6be239226be1c3a2fba2ee68dcbacaddb44c4c8', d: "m6 9 6 6 6-6" }))));
    }
    get el() { return index.getElement(this); }
};

exports.go_nav_submenu = GoNavSubmenu;
exports.go_nav_submenu_trigger = GoNavSubmenuTrigger;
