// @go-ui/core/dist/cjs/go-tab.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-348d3f4f.js');

const goTabCss = "go-tab{display:none}go-tab.active{display:block}go-tab:focus-visible{outline-offset:0.5rem}";
const GoTabStyle0 = goTabCss;

const GoTab = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.label = undefined;
        this.active = undefined;
        this.tabId = undefined;
        this.panelId = undefined;
        this.iconPosition = 'before';
        this.iconOnly = false;
    }
    async setActive(active) {
        this.active = active;
    }
    render() {
        const { panelId, tabId, active } = this;
        return (index.h(index.Host, { key: '260e446d24b1228c702c45b1ee98094be0137e17', tabindex: "0", role: "tabpanel", id: panelId, "aria-labelledby": tabId, class: { 'tab-panel': true, active } }, index.h("slot", { key: '98cf8d2a43908aa5dc6371f4f337d5029240cef2' })));
    }
    get el() { return index.getElement(this); }
};
GoTab.style = GoTabStyle0;

exports.go_tab = GoTab;
