// @go-ui/core/dist/cjs/go-to-top.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-348d3f4f.js');
const uniqueId = require('./uniqueId-93e4ecef.js');
const debounce = require('./debounce-7cb0ceca.js');

const goToTopCss = "go-to-top{--go-to-top-right:2rem;--go-to-top-bottom:2rem;display:block;position:fixed;bottom:var(--go-to-top-bottom);right:var(--go-to-top-right);pointer-events:none;z-index:var(--layer-important);opacity:0;transform:scale(0);transition:transform var(--go-duration-normal) var(--go-timing-function), opacity var(--go-duration-normal) var(--go-timing-function)}@media (prefers-reduced-motion: reduce){go-to-top{transition-duration:0s}}go-to-top.active{pointer-events:inherit;transform:scale(1);opacity:1}";
const GoToTopStyle0 = goToTopCss;

const GoToTop = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.offset = 200;
        this.active = false;
    }
    componentWillLoad() {
        // get body id or generate a new one if it doesn't exist
        this.targetId = document.body.getAttribute('id');
        if (!this.targetId) {
            this.targetId = uniqueId.uniqueId('body-');
            document.body.setAttribute('id', this.targetId);
        }
    }
    componentDidLoad() {
        window.onscroll = debounce.debounce(() => {
            this.active = window.scrollY > this.offset;
        }, 200);
    }
    render() {
        const { active } = this;
        return (index.h(index.Host, { key: '7477a98588e745f879bdbf469021d6bd3af8d6ae', class: { active } }, index.h("go-button", { key: '2ed2579eca645d6ca74471abc88d5eaea5481640', variant: "primary", flat: true, href: `#${this.targetId}`, id: "go-to-top-btn", type: "button", icon: true, round: true, "aria-labelledby": "go-to-top-btn-tooltip" }, index.h("svg", { key: '73cab3e976da71b09400d5750c491f199ac5ced7', xmlns: "http://www.w3.org/2000/svg", fill: "none", stroke: "currentColor", "stroke-linecap": "round", "stroke-linejoin": "round", "stroke-width": "2", viewBox: "0 0 24 24" }, index.h("path", { key: 'd53faba8f1b3fcd46ce2241e9d8cd82513351359', d: "m18 15-6-6-6 6" }))), index.h("go-tooltip", { key: 'd9b166056a1d6102a9a53a8d7641be33613e580f', id: "go-to-top-btn-tooltip", placement: "left", "trigger-id": "go-to-top-btn" }, "Back to top")));
    }
    get el() { return index.getElement(this); }
};
GoToTop.style = GoToTopStyle0;

exports.go_to_top = GoToTop;
