// @go-ui/core/dist/cjs/go-table-wrapper.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-348d3f4f.js');

const goTableWrapperCss = "go-table-wrapper{display:block;overflow:auto}go-table-wrapper.striped>table tr:nth-child(even){background-color:var(--go-color-table-striped-bg)}go-table-wrapper.hoverable>table tbody tr:hover,go-table-wrapper.hoverable>table tbody tr:focus{background-color:var(--go-color-table-hover-bg)}go-table-wrapper.bordered>table th,go-table-wrapper.bordered>table td,go-table-wrapper.bordered>table tr{border-width:var(--go-table-border-width-thin) 0;border-color:var(--go-color-table-border);border-style:solid}go-table-wrapper.bordered>table thead th{border-width:0 var(--go-table-border-width-thin)}go-table-wrapper.bordered>table tbody td,go-table-wrapper.bordered>table tbody th{border-width:0 var(--go-table-border-width-thin)}";
const GoTableWrapperStyle0 = goTableWrapperCss;

const GoTableWrapper = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.striped = false;
        this.bordered = false;
        this.hoverable = false;
    }
    // Store attributes inherited from the host element
    componentWillLoad() { }
    render() {
        const { striped, bordered, hoverable } = this;
        return (index.h(index.Host, { key: 'c707d76b8c23c8d33cf1d059274ef2085a1a0568', class: {
                striped,
                bordered,
                hoverable,
            } }, index.h("slot", { key: '0e3b403c031b5ed09305ce96c98d4d5ebbcfc1b8' })));
    }
    get el() { return index.getElement(this); }
};
GoTableWrapper.style = GoTableWrapperStyle0;

exports.go_table_wrapper = GoTableWrapper;
