// @go-ui/core/dist/cjs/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const form = require('./form-6e13a9ec.js');
const dom = require('./dom-7cff1398.js');
const goUiMdPlugin = require('./go-ui-md-plugin-476649f1.js');
const helper = require('./helper-c1931cd5.js');
require('./kebabCase-f98d4407.js');
require('./uniqueId-93e4ecef.js');
require('./index-c48c451e.js');



exports.fieldPropNames = form.fieldPropNames;
exports.fieldSlotNames = form.fieldSlotNames;
exports.inheritNonFieldAttrs = form.inheritNonFieldAttrs;
exports.loadFieldProps = form.loadFieldProps;
exports.loadFieldSlots = form.loadFieldSlots;
exports.AUTO_FOCUS_TIMEOUT = dom.AUTO_FOCUS_TIMEOUT;
exports.fadeOutRemove = dom.fadeOutRemove;
exports.focusFirstWithin = dom.focusFirstWithin;
exports.focusLastWithin = dom.focusLastWithin;
exports.getFirstFocusableChild = dom.getFirstFocusableChild;
exports.getFocusableChildren = dom.getFocusableChildren;
exports.getLastFocusableChild = dom.getLastFocusableChild;
exports.moveEl = dom.moveEl;
exports.onAnimationEnd = dom.onAnimationEnd;
exports.onClickOutside = dom.onClickOutside;
exports.onEscape = dom.onEscape;
exports.tabIndexes = dom.tabIndexes;
exports.trapFocus = dom.trapFocus;
exports.goUiPlugin = goUiMdPlugin.goUiPlugin;
exports.$attrs = helper.$attrs;
exports.extractId = helper.extractId;
exports.hasShadowDom = helper.hasShadowDom;
exports.hasSlot = helper.hasSlot;
exports.inheritAttributes = helper.inheritAttributes;
exports.inheritComponentAttrs = helper.inheritComponentAttrs;
exports.initIdProps = helper.initIdProps;
exports.isDarkMode = helper.isDarkMode;
exports.mm = helper.mm;
exports.parseJsonProp = helper.parseJsonProp;
exports.prefersDark = helper.prefersDark;
exports.prefersReducedMotion = helper.prefersReducedMotion;
exports.selectDirectChildren = helper.selectDirectChildren;
exports.warning = helper.warning;
exports.watchDevice = helper.watchDevice;
